// swagger-ts-generator/test-gen/app/models/webapi-v3/address.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';


export interface IAddress {
    street?: string;
    city?: string;
    state?: string;
    zip?: string;
}


export class Address extends BaseModel implements IAddress  {

    static STREET_FIELD_NAME = 'street';
    static CITY_FIELD_NAME = 'city';
    static STATE_FIELD_NAME = 'state';
    static ZIP_FIELD_NAME = 'zip';

    street: string;
    city: string;
    state: string;
    zip: string;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    
    */
    constructor(values?: Partial<IAddress>) {
        super();

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<IAddress>): void {
        if (values) {
            const rawValues = values
            this.street = this.getValue<string>(rawValues, Address.STREET_FIELD_NAME);
            this.city = this.getValue<string>(rawValues, Address.CITY_FIELD_NAME);
            this.state = this.getValue<string>(rawValues, Address.STATE_FIELD_NAME);
            this.zip = this.getValue<string>(rawValues, Address.ZIP_FIELD_NAME);
        }
    }

}
