// swagger-ts-generator/test-gen/app/models/webapi-v3/api-response.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';


export interface IApiResponse {
    code?: number;
    type?: string;
    message?: string;
}


export class ApiResponse extends BaseModel implements IApiResponse  {

    static CODE_FIELD_NAME = 'code';
    static TYPE_FIELD_NAME = 'type';
    static MESSAGE_FIELD_NAME = 'message';

    code: number;
    type: string;
    message: string;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    
    */
    constructor(values?: Partial<IApiResponse>) {
        super();

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<IApiResponse>): void {
        if (values) {
            const rawValues = values
            this.code = this.getValue<number>(rawValues, ApiResponse.CODE_FIELD_NAME);
            this.type = this.getValue<string>(rawValues, ApiResponse.TYPE_FIELD_NAME);
            this.message = this.getValue<string>(rawValues, ApiResponse.MESSAGE_FIELD_NAME);
        }
    }

}
