// swagger-ts-generator/test-gen/app/models/webapi/auth-0-config.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';


export interface IAuth0Config {
    audience?: string;
    clientID?: string;
    domain?: string;
    loginEndpoint?: string;
    responseType?: string;
    scope?: string;
}


export class Auth0Config extends BaseModel implements IAuth0Config  {

    static AUDIENCE_FIELD_NAME = 'audience';
    static CLIENT_ID_FIELD_NAME = 'clientID';
    static DOMAIN_FIELD_NAME = 'domain';
    static LOGIN_ENDPOINT_FIELD_NAME = 'loginEndpoint';
    static RESPONSE_TYPE_FIELD_NAME = 'responseType';
    static SCOPE_FIELD_NAME = 'scope';

    audience: string;
    clientID: string;
    domain: string;
    loginEndpoint: string;
    responseType: string;
    scope: string;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    */
    constructor(values?: Partial<IAuth0Config>) {
        super();

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<IAuth0Config>): void {
        if (values) {
            
            const rawValues = values
            this.audience = this.getValue<string>(rawValues, Auth0Config.AUDIENCE_FIELD_NAME);
            this.clientID = this.getValue<string>(rawValues, Auth0Config.CLIENT_ID_FIELD_NAME);
            this.domain = this.getValue<string>(rawValues, Auth0Config.DOMAIN_FIELD_NAME);
            this.loginEndpoint = this.getValue<string>(rawValues, Auth0Config.LOGIN_ENDPOINT_FIELD_NAME);
            this.responseType = this.getValue<string>(rawValues, Auth0Config.RESPONSE_TYPE_FIELD_NAME);
            this.scope = this.getValue<string>(rawValues, Auth0Config.SCOPE_FIELD_NAME);
        }
    }
}
