// swagger-ts-generator/test-gen/app/models/webapi/azure-sas-info.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';


export interface IAzureSasInfo {
    accountName?: string;
    alreadyExist?: boolean;
    containerName?: string;
    createdAt?: Date;
    endpointSuffix?: string;
    fileName?: string;
    lastModifiedAt?: Date;
    sasToken?: string;
}


export class AzureSasInfo extends BaseModel implements IAzureSasInfo  {

    static ACCOUNT_NAME_FIELD_NAME = 'accountName';
    static ALREADY_EXIST_FIELD_NAME = 'alreadyExist';
    static CONTAINER_NAME_FIELD_NAME = 'containerName';
    static CREATED_AT_FIELD_NAME = 'createdAt';
    static ENDPOINT_SUFFIX_FIELD_NAME = 'endpointSuffix';
    static FILE_NAME_FIELD_NAME = 'fileName';
    static LAST_MODIFIED_AT_FIELD_NAME = 'lastModifiedAt';
    static SAS_TOKEN_FIELD_NAME = 'sasToken';

    accountName: string;
    alreadyExist: boolean;
    containerName: string;
    createdAt: Date;
    endpointSuffix: string;
    fileName: string;
    lastModifiedAt: Date;
    sasToken: string;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    */
    constructor(values?: Partial<IAzureSasInfo>) {
        super();

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<IAzureSasInfo>): void {
        if (values) {
            
            const rawValues = values
            this.accountName = this.getValue<string>(rawValues, AzureSasInfo.ACCOUNT_NAME_FIELD_NAME);
            this.alreadyExist = this.getValue<boolean>(rawValues, AzureSasInfo.ALREADY_EXIST_FIELD_NAME);
            this.containerName = this.getValue<string>(rawValues, AzureSasInfo.CONTAINER_NAME_FIELD_NAME);
            this.createdAt = this.getValue<Date>(rawValues, AzureSasInfo.CREATED_AT_FIELD_NAME);
            this.endpointSuffix = this.getValue<string>(rawValues, AzureSasInfo.ENDPOINT_SUFFIX_FIELD_NAME);
            this.fileName = this.getValue<string>(rawValues, AzureSasInfo.FILE_NAME_FIELD_NAME);
            this.lastModifiedAt = this.getValue<Date>(rawValues, AzureSasInfo.LAST_MODIFIED_AT_FIELD_NAME);
            this.sasToken = this.getValue<string>(rawValues, AzureSasInfo.SAS_TOKEN_FIELD_NAME);
        }
    }
}
