// swagger-ts-generator/test-gen/app/models/webapi/azure-file.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';


export interface IAzureFile {
    containerName?: string;
    createdAt?: Date;
    fileHash?: number;
    fileName?: string;
    folderName?: string;
    id?: string;
    lastModifiedAt?: Date;
    originalFileName?: string;
    typeSelector?: string;
    version?: number;
}


export class AzureFile extends BaseModel implements IAzureFile  {
    static TYPE_SELECTOR = 'AzureFile';

    static CONTAINER_NAME_FIELD_NAME = 'containerName';
    static CREATED_AT_FIELD_NAME = 'createdAt';
    static FILE_HASH_FIELD_NAME = 'fileHash';
    static FILE_NAME_FIELD_NAME = 'fileName';
    static FOLDER_NAME_FIELD_NAME = 'folderName';
    static ID_FIELD_NAME = 'id';
    static LAST_MODIFIED_AT_FIELD_NAME = 'lastModifiedAt';
    static ORIGINAL_FILE_NAME_FIELD_NAME = 'originalFileName';
    static TYPE_SELECTOR_FIELD_NAME = 'typeSelector';
    static VERSION_FIELD_NAME = 'version';

    containerName: string;
    createdAt: Date;
    fileHash: number;
    fileName: string;
    folderName: string;
    id: string;
    lastModifiedAt: Date;
    originalFileName: string;
    typeSelector: string;
    version: number;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    */
    constructor(values?: Partial<IAzureFile>) {
        super();

        this.typeSelector = AzureFile.TYPE_SELECTOR;

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<IAzureFile>): void {
        if (values) {
            
            const rawValues = values
            this.containerName = this.getValue<string>(rawValues, AzureFile.CONTAINER_NAME_FIELD_NAME);
            this.createdAt = this.getValue<Date>(rawValues, AzureFile.CREATED_AT_FIELD_NAME);
            this.fileHash = this.getValue<number>(rawValues, AzureFile.FILE_HASH_FIELD_NAME);
            this.fileName = this.getValue<string>(rawValues, AzureFile.FILE_NAME_FIELD_NAME);
            this.folderName = this.getValue<string>(rawValues, AzureFile.FOLDER_NAME_FIELD_NAME);
            this.id = this.getValue<string>(rawValues, AzureFile.ID_FIELD_NAME);
            this.lastModifiedAt = this.getValue<Date>(rawValues, AzureFile.LAST_MODIFIED_AT_FIELD_NAME);
            this.originalFileName = this.getValue<string>(rawValues, AzureFile.ORIGINAL_FILE_NAME_FIELD_NAME);
            this.typeSelector = this.getValue<string>(rawValues, AzureFile.TYPE_SELECTOR_FIELD_NAME);
            this.version = this.getValue<number>(rawValues, AzureFile.VERSION_FIELD_NAME);
        }
    }
}
