// swagger-ts-generator/test-gen/app/models/webapi/bom-reinsurance-instalment.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';

import { IMoney, Money } from './money.model';

export interface IBomReinsuranceInstalment {
    brokerageInstalmentAmount?: IMoney;
    brokerageInstalmentId?: string;
    instalmentDate?: Date;
    premiumInstalmentAmount?: IMoney;
    premiumInstalmentId?: string;
}


export class BomReinsuranceInstalment extends BaseModel implements IBomReinsuranceInstalment  {

    static BROKERAGE_INSTALMENT_AMOUNT_FIELD_NAME = 'brokerageInstalmentAmount';
    static BROKERAGE_INSTALMENT_ID_FIELD_NAME = 'brokerageInstalmentId';
    static INSTALMENT_DATE_FIELD_NAME = 'instalmentDate';
    static PREMIUM_INSTALMENT_AMOUNT_FIELD_NAME = 'premiumInstalmentAmount';
    static PREMIUM_INSTALMENT_ID_FIELD_NAME = 'premiumInstalmentId';

    brokerageInstalmentAmount: Money;
    brokerageInstalmentId: string;
    instalmentDate: Date;
    premiumInstalmentAmount: Money;
    premiumInstalmentId: string;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    */
    constructor(values?: Partial<IBomReinsuranceInstalment>) {
        super();
        this.brokerageInstalmentAmount = new Money(); 
        this.premiumInstalmentAmount = new Money(); 

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<IBomReinsuranceInstalment>): void {
        if (values) {
            
            const rawValues = values
            this.brokerageInstalmentAmount.setValues(rawValues.brokerageInstalmentAmount);
            this.brokerageInstalmentId = this.getValue<string>(rawValues, BomReinsuranceInstalment.BROKERAGE_INSTALMENT_ID_FIELD_NAME);
            this.instalmentDate = this.getValue<Date>(rawValues, BomReinsuranceInstalment.INSTALMENT_DATE_FIELD_NAME);
            this.premiumInstalmentAmount.setValues(rawValues.premiumInstalmentAmount);
            this.premiumInstalmentId = this.getValue<string>(rawValues, BomReinsuranceInstalment.PREMIUM_INSTALMENT_ID_FIELD_NAME);
        }
    }
}
