// swagger-ts-generator/test-gen/app/models/webapi/bom-corda-tx-execution-state-notification.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';

import { Codex1EntityTypeSelector } from './enums';
import { BomTransactionStatus } from './enums';
import { BomWebSocketEventType } from './enums';

export interface IBomCordaTxExecutionStateNotification {
    entityTypeSelector?: Codex1EntityTypeSelector;
    stateId?: string;
    status?: BomTransactionStatus;
    webSocketEventType?: BomWebSocketEventType;
}


export class BomCordaTxExecutionStateNotification extends BaseModel implements IBomCordaTxExecutionStateNotification  {

    static ENTITY_TYPE_SELECTOR_FIELD_NAME = 'entityTypeSelector';
    static STATE_ID_FIELD_NAME = 'stateId';
    static STATUS_FIELD_NAME = 'status';
    static WEB_SOCKET_EVENT_TYPE_FIELD_NAME = 'webSocketEventType';

    entityTypeSelector: Codex1EntityTypeSelector;
    stateId: string;
    status: BomTransactionStatus;
    webSocketEventType: BomWebSocketEventType;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    */
    constructor(values?: Partial<IBomCordaTxExecutionStateNotification>) {
        super();

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<IBomCordaTxExecutionStateNotification>): void {
        if (values) {
            
            const rawValues = values
            this.entityTypeSelector = this.getValue<Codex1EntityTypeSelector>(rawValues, BomCordaTxExecutionStateNotification.ENTITY_TYPE_SELECTOR_FIELD_NAME);
            this.stateId = this.getValue<string>(rawValues, BomCordaTxExecutionStateNotification.STATE_ID_FIELD_NAME);
            this.status = this.getValue<BomTransactionStatus>(rawValues, BomCordaTxExecutionStateNotification.STATUS_FIELD_NAME);
            this.webSocketEventType = this.getValue<BomWebSocketEventType>(rawValues, BomCordaTxExecutionStateNotification.WEB_SOCKET_EVENT_TYPE_FIELD_NAME);
        }
    }
}
