// swagger-ts-generator/test-gen/app/models/webapi/bom-quota-share-loss-advice-entry.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';

import { IMoney, Money } from './money.model';

export interface IBomQuotaShareLossAdviceEntry {
    lossAdviceId?: string;
    lossAmount?: IMoney;
    sections?: Array<string>;
}


export class BomQuotaShareLossAdviceEntry extends BaseModel implements IBomQuotaShareLossAdviceEntry  {

    static LOSS_ADVICE_ID_FIELD_NAME = 'lossAdviceId';
    static LOSS_AMOUNT_FIELD_NAME = 'lossAmount';
    static SECTIONS_FIELD_NAME = 'sections';

    lossAdviceId: string;
    lossAmount: Money;
    sections: Array<string>;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    */
    constructor(values?: Partial<IBomQuotaShareLossAdviceEntry>) {
        super();
        this.lossAmount = new Money(); 
        this.sections = new Array<string>(); 

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<IBomQuotaShareLossAdviceEntry>): void {
        if (values) {
            
            const rawValues = values
            this.lossAdviceId = this.getValue<string>(rawValues, BomQuotaShareLossAdviceEntry.LOSS_ADVICE_ID_FIELD_NAME);
            this.lossAmount.setValues(rawValues.lossAmount);
            this.fillModelArray<string>(this, BomQuotaShareLossAdviceEntry.SECTIONS_FIELD_NAME, rawValues.sections);
        }
    }
}
