// swagger-ts-generator/test-gen/app/models/webapi/bom-reinsurance-contract.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';

import { IBomReinsuranceClause, BomReinsuranceClause } from './bom-reinsurance-clause.model';
import { IContractWording, ContractWording } from './contract-wording.model';
import { IBomDocument, BomDocument } from './bom-document.model';
import { IBomBoundedPeriod, BomBoundedPeriod } from './bom-bounded-period.model';
import { IBomCatXolReinsuranceLayer, BomCatXolReinsuranceLayer } from './bom-cat-xol-reinsurance-layer.model';
import { IBomCatXolReinsuranceSection, BomCatXolReinsuranceSection } from './bom-cat-xol-reinsurance-section.model';
import { IBomQuotaShareLayer, BomQuotaShareLayer } from './bom-quota-share-layer.model';
import { IBomQuotaShareSection, BomQuotaShareSection } from './bom-quota-share-section.model';
import { IBomParticipantInRole, BomParticipantInRole } from './bom-participant-in-role.model';
import { IBomReinsuranceWrittenAndSignedAudit, BomReinsuranceWrittenAndSignedAudit } from './bom-reinsurance-written-and-signed-audit.model';
import { ContractChangeType } from './enums';
import { BomReinsuranceContractType } from './enums';
import { BomReinsuranceContractProcessType } from './enums';
import { BomReinsuranceBasis } from './enums';
import { BomReinsuranceContractStatus } from './enums';
import { BomTransactionStatus } from './enums';

export interface IBomReinsuranceContract {
    clauses?: Array<IBomReinsuranceClause>;
    contractChangeType?: ContractChangeType;
    contractType?: BomReinsuranceContractType;
    contractWording?: IContractWording;
    createdAt?: Date;
    description?: string;
    documents?: Array<IBomDocument>;
    effectivePeriod?: IBomBoundedPeriod;
    endorsedContractId?: string;
    followUpDueAt?: Date;
    id?: string;
    insuranceContractProcessId?: string;
    insuranceContractProcessType?: BomReinsuranceContractProcessType;
    insuranceServiceRequestId?: string;
    isInTransientStore?: boolean;
    lastModifiedAt?: Date;
    layers?: Array<IBomCatXolReinsuranceLayer | IBomCatXolReinsuranceSection | IBomQuotaShareLayer | IBomQuotaShareSection>;
    masterContractId?: string;
    myCompanyParticipantInRole?: IBomParticipantInRole;
    name: string;
    negotiationCount?: number;
    nextActionExpectedFrom?: IBomParticipantInRole;
    participants?: Array<IBomParticipantInRole>;
    reinsuranceBasis?: BomReinsuranceBasis;
    signedAudits?: Array<IBomReinsuranceWrittenAndSignedAudit>;
    status?: BomReinsuranceContractStatus;
    subjectivities?: Array<IBomReinsuranceClause>;
    targetStatus?: BomReinsuranceContractStatus;
    transactionStatus?: BomTransactionStatus;
    treatyId?: string;
    version?: number;
    writtenAudits?: Array<IBomReinsuranceWrittenAndSignedAudit>;
}


export class BomReinsuranceContract extends BaseModel implements IBomReinsuranceContract  {

    static CLAUSES_FIELD_NAME = 'clauses';
    static CONTRACT_CHANGE_TYPE_FIELD_NAME = 'contractChangeType';
    static CONTRACT_TYPE_FIELD_NAME = 'contractType';
    static CONTRACT_WORDING_FIELD_NAME = 'contractWording';
    static CREATED_AT_FIELD_NAME = 'createdAt';
    static DESCRIPTION_FIELD_NAME = 'description';
    static DOCUMENTS_FIELD_NAME = 'documents';
    static EFFECTIVE_PERIOD_FIELD_NAME = 'effectivePeriod';
    static ENDORSED_CONTRACT_ID_FIELD_NAME = 'endorsedContractId';
    static FOLLOW_UP_DUE_AT_FIELD_NAME = 'followUpDueAt';
    static ID_FIELD_NAME = 'id';
    static INSURANCE_CONTRACT_PROCESS_ID_FIELD_NAME = 'insuranceContractProcessId';
    static INSURANCE_CONTRACT_PROCESS_TYPE_FIELD_NAME = 'insuranceContractProcessType';
    static INSURANCE_SERVICE_REQUEST_ID_FIELD_NAME = 'insuranceServiceRequestId';
    static IS_IN_TRANSIENT_STORE_FIELD_NAME = 'isInTransientStore';
    static LAST_MODIFIED_AT_FIELD_NAME = 'lastModifiedAt';
    static LAYERS_FIELD_NAME = 'layers';
    static MASTER_CONTRACT_ID_FIELD_NAME = 'masterContractId';
    static MY_COMPANY_PARTICIPANT_IN_ROLE_FIELD_NAME = 'myCompanyParticipantInRole';
    static NAME_FIELD_NAME = 'name';
    static NEGOTIATION_COUNT_FIELD_NAME = 'negotiationCount';
    static NEXT_ACTION_EXPECTED_FROM_FIELD_NAME = 'nextActionExpectedFrom';
    static PARTICIPANTS_FIELD_NAME = 'participants';
    static REINSURANCE_BASIS_FIELD_NAME = 'reinsuranceBasis';
    static SIGNED_AUDITS_FIELD_NAME = 'signedAudits';
    static STATUS_FIELD_NAME = 'status';
    static SUBJECTIVITIES_FIELD_NAME = 'subjectivities';
    static TARGET_STATUS_FIELD_NAME = 'targetStatus';
    static TRANSACTION_STATUS_FIELD_NAME = 'transactionStatus';
    static TREATY_ID_FIELD_NAME = 'treatyId';
    static VERSION_FIELD_NAME = 'version';
    static WRITTEN_AUDITS_FIELD_NAME = 'writtenAudits';

    clauses: Array<BomReinsuranceClause>;
    contractChangeType: ContractChangeType;
    contractType: BomReinsuranceContractType;
    contractWording: ContractWording;
    createdAt: Date;
    description: string;
    documents: Array<BomDocument>;
    effectivePeriod: BomBoundedPeriod;
    endorsedContractId: string;
    followUpDueAt: Date;
    id: string;
    insuranceContractProcessId: string;
    insuranceContractProcessType: BomReinsuranceContractProcessType;
    insuranceServiceRequestId: string;
    isInTransientStore: boolean;
    lastModifiedAt: Date;
    layers: Array<BomCatXolReinsuranceLayer | BomCatXolReinsuranceSection | BomQuotaShareLayer | BomQuotaShareSection>;
    masterContractId: string;
    myCompanyParticipantInRole: BomParticipantInRole;
    name: string;
    negotiationCount: number;
    nextActionExpectedFrom: BomParticipantInRole;
    participants: Array<BomParticipantInRole>;
    reinsuranceBasis: BomReinsuranceBasis;
    signedAudits: Array<BomReinsuranceWrittenAndSignedAudit>;
    status: BomReinsuranceContractStatus;
    subjectivities: Array<BomReinsuranceClause>;
    targetStatus: BomReinsuranceContractStatus;
    transactionStatus: BomTransactionStatus;
    treatyId: string;
    version: number;
    writtenAudits: Array<BomReinsuranceWrittenAndSignedAudit>;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    */
    constructor(values?: Partial<IBomReinsuranceContract>) {
        super();
        this.clauses = new Array<BomReinsuranceClause>(); 
        this.contractWording = new ContractWording(); 
        this.documents = new Array<BomDocument>(); 
        this.effectivePeriod = new BomBoundedPeriod(); 
        this.layers = new Array<BomCatXolReinsuranceLayer | BomCatXolReinsuranceSection | BomQuotaShareLayer | BomQuotaShareSection>(); 
        this.myCompanyParticipantInRole = new BomParticipantInRole(); 
        this.nextActionExpectedFrom = new BomParticipantInRole(); 
        this.participants = new Array<BomParticipantInRole>(); 
        this.signedAudits = new Array<BomReinsuranceWrittenAndSignedAudit>(); 
        this.subjectivities = new Array<BomReinsuranceClause>(); 
        this.writtenAudits = new Array<BomReinsuranceWrittenAndSignedAudit>(); 

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<IBomReinsuranceContract>): void {
        if (values) {
            
            const rawValues = values
            this.fillModelArray<IBomReinsuranceClause>(this, BomReinsuranceContract.CLAUSES_FIELD_NAME, rawValues.clauses, BomReinsuranceClause, SubTypeFactory.createSubTypeInstance);
            this.contractChangeType = this.getValue<ContractChangeType>(rawValues, BomReinsuranceContract.CONTRACT_CHANGE_TYPE_FIELD_NAME);
            this.contractType = this.getValue<BomReinsuranceContractType>(rawValues, BomReinsuranceContract.CONTRACT_TYPE_FIELD_NAME);
            this.contractWording.setValues(rawValues.contractWording);
            this.createdAt = this.getValue<Date>(rawValues, BomReinsuranceContract.CREATED_AT_FIELD_NAME);
            this.description = this.getValue<string>(rawValues, BomReinsuranceContract.DESCRIPTION_FIELD_NAME);
            this.fillModelArray<IBomDocument>(this, BomReinsuranceContract.DOCUMENTS_FIELD_NAME, rawValues.documents, BomDocument, SubTypeFactory.createSubTypeInstance);
            this.effectivePeriod.setValues(rawValues.effectivePeriod);
            this.endorsedContractId = this.getValue<string>(rawValues, BomReinsuranceContract.ENDORSED_CONTRACT_ID_FIELD_NAME);
            this.followUpDueAt = this.getValue<Date>(rawValues, BomReinsuranceContract.FOLLOW_UP_DUE_AT_FIELD_NAME);
            this.id = this.getValue<string>(rawValues, BomReinsuranceContract.ID_FIELD_NAME);
            this.insuranceContractProcessId = this.getValue<string>(rawValues, BomReinsuranceContract.INSURANCE_CONTRACT_PROCESS_ID_FIELD_NAME);
            this.insuranceContractProcessType = this.getValue<BomReinsuranceContractProcessType>(rawValues, BomReinsuranceContract.INSURANCE_CONTRACT_PROCESS_TYPE_FIELD_NAME);
            this.insuranceServiceRequestId = this.getValue<string>(rawValues, BomReinsuranceContract.INSURANCE_SERVICE_REQUEST_ID_FIELD_NAME);
            this.isInTransientStore = this.getValue<boolean>(rawValues, BomReinsuranceContract.IS_IN_TRANSIENT_STORE_FIELD_NAME);
            this.lastModifiedAt = this.getValue<Date>(rawValues, BomReinsuranceContract.LAST_MODIFIED_AT_FIELD_NAME);
            this.fillModelArray<IBomCatXolReinsuranceLayer | IBomCatXolReinsuranceSection | IBomQuotaShareLayer | IBomQuotaShareSection>(this, BomReinsuranceContract.LAYERS_FIELD_NAME, rawValues.layers, BomCatXolReinsuranceLayer, SubTypeFactory.createSubTypeInstance);
            this.masterContractId = this.getValue<string>(rawValues, BomReinsuranceContract.MASTER_CONTRACT_ID_FIELD_NAME);
            this.myCompanyParticipantInRole.setValues(rawValues.myCompanyParticipantInRole);
            this.name = this.getValue<string>(rawValues, BomReinsuranceContract.NAME_FIELD_NAME);
            this.negotiationCount = this.getValue<number>(rawValues, BomReinsuranceContract.NEGOTIATION_COUNT_FIELD_NAME);
            this.nextActionExpectedFrom.setValues(rawValues.nextActionExpectedFrom);
            this.fillModelArray<IBomParticipantInRole>(this, BomReinsuranceContract.PARTICIPANTS_FIELD_NAME, rawValues.participants, BomParticipantInRole, SubTypeFactory.createSubTypeInstance);
            this.reinsuranceBasis = this.getValue<BomReinsuranceBasis>(rawValues, BomReinsuranceContract.REINSURANCE_BASIS_FIELD_NAME);
            this.fillModelArray<IBomReinsuranceWrittenAndSignedAudit>(this, BomReinsuranceContract.SIGNED_AUDITS_FIELD_NAME, rawValues.signedAudits, BomReinsuranceWrittenAndSignedAudit, SubTypeFactory.createSubTypeInstance);
            this.status = this.getValue<BomReinsuranceContractStatus>(rawValues, BomReinsuranceContract.STATUS_FIELD_NAME);
            this.fillModelArray<IBomReinsuranceClause>(this, BomReinsuranceContract.SUBJECTIVITIES_FIELD_NAME, rawValues.subjectivities, BomReinsuranceClause, SubTypeFactory.createSubTypeInstance);
            this.targetStatus = this.getValue<BomReinsuranceContractStatus>(rawValues, BomReinsuranceContract.TARGET_STATUS_FIELD_NAME);
            this.transactionStatus = this.getValue<BomTransactionStatus>(rawValues, BomReinsuranceContract.TRANSACTION_STATUS_FIELD_NAME);
            this.treatyId = this.getValue<string>(rawValues, BomReinsuranceContract.TREATY_ID_FIELD_NAME);
            this.version = this.getValue<number>(rawValues, BomReinsuranceContract.VERSION_FIELD_NAME);
            this.fillModelArray<IBomReinsuranceWrittenAndSignedAudit>(this, BomReinsuranceContract.WRITTEN_AUDITS_FIELD_NAME, rawValues.writtenAudits, BomReinsuranceWrittenAndSignedAudit, SubTypeFactory.createSubTypeInstance);
        }
    }
}
