// swagger-ts-generator/test-gen/app/models/webapi/bom-reinsurance-clause.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';

import { ClauseType } from './enums';

export interface IBomReinsuranceClause {
    description: string;
    id?: string;
    name: string;
    number?: number;
    type?: ClauseType;
}


export class BomReinsuranceClause extends BaseModel implements IBomReinsuranceClause  {

    static DESCRIPTION_FIELD_NAME = 'description';
    static ID_FIELD_NAME = 'id';
    static NAME_FIELD_NAME = 'name';
    static NUMBER_FIELD_NAME = 'number';
    static TYPE_FIELD_NAME = 'type';

    description: string;
    id: string;
    name: string;
    number: number;
    type: ClauseType;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    */
    constructor(values?: Partial<IBomReinsuranceClause>) {
        super();

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<IBomReinsuranceClause>): void {
        if (values) {
            
            const rawValues = values
            this.description = this.getValue<string>(rawValues, BomReinsuranceClause.DESCRIPTION_FIELD_NAME);
            this.id = this.getValue<string>(rawValues, BomReinsuranceClause.ID_FIELD_NAME);
            this.name = this.getValue<string>(rawValues, BomReinsuranceClause.NAME_FIELD_NAME);
            this.number = this.getValue<number>(rawValues, BomReinsuranceClause.NUMBER_FIELD_NAME);
            this.type = this.getValue<ClauseType>(rawValues, BomReinsuranceClause.TYPE_FIELD_NAME);
        }
    }
}
