// swagger-ts-generator/test-gen/app/models/webapi/bom-quota-share-line.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';


export interface IBomQuotaShareLine {
    minimumLine?: string;
    sectionId?: string;
    signedLine?: string;
    writtenLine?: string;
}


export class BomQuotaShareLine extends BaseModel implements IBomQuotaShareLine  {

    static MINIMUM_LINE_FIELD_NAME = 'minimumLine';
    static SECTION_ID_FIELD_NAME = 'sectionId';
    static SIGNED_LINE_FIELD_NAME = 'signedLine';
    static WRITTEN_LINE_FIELD_NAME = 'writtenLine';

    minimumLine: string;
    sectionId: string;
    signedLine: string;
    writtenLine: string;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    */
    constructor(values?: Partial<IBomQuotaShareLine>) {
        super();

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<IBomQuotaShareLine>): void {
        if (values) {
            
            const rawValues = values
            this.minimumLine = this.getValue<string>(rawValues, BomQuotaShareLine.MINIMUM_LINE_FIELD_NAME);
            this.sectionId = this.getValue<string>(rawValues, BomQuotaShareLine.SECTION_ID_FIELD_NAME);
            this.signedLine = this.getValue<string>(rawValues, BomQuotaShareLine.SIGNED_LINE_FIELD_NAME);
            this.writtenLine = this.getValue<string>(rawValues, BomQuotaShareLine.WRITTEN_LINE_FIELD_NAME);
        }
    }
}
