// swagger-ts-generator/test-gen/app/models/webapi/bom-quota-share-layer.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';

import { IBomBaseReinsuranceSection, BomBaseReinsuranceSection } from './bom-base-reinsurance-section.model';
import { IBomReinsurancePortfolio, BomReinsurancePortfolio } from './bom-reinsurance-portfolio.model';
import { IBomQuotaShareCashLoss, BomQuotaShareCashLoss } from './bom-quota-share-cash-loss.model';
import { IBomQuotaShareLine, BomQuotaShareLine } from './bom-quota-share-line.model';
import { IBomQuotaShareLossAdvice, BomQuotaShareLossAdvice } from './bom-quota-share-loss-advice.model';
import { IBomFlatOrRateBrokerage, BomFlatOrRateBrokerage } from './bom-flat-or-rate-brokerage.model';
import { IBomCedingCommission, BomCedingCommission } from './bom-ceding-commission.model';
import { IBomQuotaShareSection, BomQuotaShareSection } from './bom-quota-share-section.model';
import { BomApplicableLawCode } from './enums';

export interface IBomQuotaShareLayer extends IBomBaseReinsuranceSection{
    portfolios?: Array<IBomReinsurancePortfolio>;
    cashLoss?: IBomQuotaShareCashLoss;
    lines?: Array<IBomQuotaShareLine>;
    lossAdvice?: IBomQuotaShareLossAdvice;
    applicableLaw?: BomApplicableLawCode;
    brokerage?: IBomFlatOrRateBrokerage;
    cedingCommission?: IBomCedingCommission;
    sections?: Array<IBomQuotaShareSection>;
}


export class BomQuotaShareLayer extends BomBaseReinsuranceSection implements IBomQuotaShareLayer  {
    static TYPE_SELECTOR = 'BomQuotaShareLayer';

    static PORTFOLIOS_FIELD_NAME = 'portfolios';
    static CASH_LOSS_FIELD_NAME = 'cashLoss';
    static LINES_FIELD_NAME = 'lines';
    static LOSS_ADVICE_FIELD_NAME = 'lossAdvice';
    static APPLICABLE_LAW_FIELD_NAME = 'applicableLaw';
    static BROKERAGE_FIELD_NAME = 'brokerage';
    static CEDING_COMMISSION_FIELD_NAME = 'cedingCommission';
    static SECTIONS_FIELD_NAME = 'sections';

    portfolios: Array<BomReinsurancePortfolio>;
    cashLoss: BomQuotaShareCashLoss;
    lines: Array<BomQuotaShareLine>;
    lossAdvice: BomQuotaShareLossAdvice;
    applicableLaw: BomApplicableLawCode;
    brokerage: BomFlatOrRateBrokerage;
    cedingCommission: BomCedingCommission;
    sections: Array<BomQuotaShareSection>;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    */
    constructor(values?: Partial<IBomQuotaShareLayer>) {
        super();

        this.typeSelector = BomQuotaShareLayer.TYPE_SELECTOR;
        this.portfolios = new Array<BomReinsurancePortfolio>(); 
        this.cashLoss = new BomQuotaShareCashLoss(); 
        this.lines = new Array<BomQuotaShareLine>(); 
        this.lossAdvice = new BomQuotaShareLossAdvice(); 
        this.brokerage = new BomFlatOrRateBrokerage(); 
        this.cedingCommission = new BomCedingCommission(); 
        this.sections = new Array<BomQuotaShareSection>(); 

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<IBomQuotaShareLayer>): void {
        if (values) {
            super.setValues(values)
            const rawValues = values
            this.fillModelArray<IBomReinsurancePortfolio>(this, BomQuotaShareLayer.PORTFOLIOS_FIELD_NAME, rawValues.portfolios, BomReinsurancePortfolio, SubTypeFactory.createSubTypeInstance);
            this.cashLoss.setValues(rawValues.cashLoss);
            this.fillModelArray<IBomQuotaShareLine>(this, BomQuotaShareLayer.LINES_FIELD_NAME, rawValues.lines, BomQuotaShareLine, SubTypeFactory.createSubTypeInstance);
            this.lossAdvice.setValues(rawValues.lossAdvice);
            this.applicableLaw = this.getValue<BomApplicableLawCode>(rawValues, BomQuotaShareLayer.APPLICABLE_LAW_FIELD_NAME);
            this.brokerage.setValues(rawValues.brokerage);
            this.cedingCommission.setValues(rawValues.cedingCommission);
            this.fillModelArray<IBomQuotaShareSection>(this, BomQuotaShareLayer.SECTIONS_FIELD_NAME, rawValues.sections, BomQuotaShareSection, SubTypeFactory.createSubTypeInstance);
        }
    }
}
