// swagger-ts-generator/test-gen/app/models/webapi/bom-telephone-number.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';

import { TelecomType } from './enums';

export interface IBomTelephoneNumber {
    countryCode?: string;
    extension?: string;
    number?: string;
    type?: TelecomType;
}


export class BomTelephoneNumber extends BaseModel implements IBomTelephoneNumber  {

    static COUNTRY_CODE_FIELD_NAME = 'countryCode';
    static EXTENSION_FIELD_NAME = 'extension';
    static NUMBER_FIELD_NAME = 'number';
    static TYPE_FIELD_NAME = 'type';

    countryCode: string;
    extension: string;
    number: string;
    type: TelecomType;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    */
    constructor(values?: Partial<IBomTelephoneNumber>) {
        super();

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<IBomTelephoneNumber>): void {
        if (values) {
            
            const rawValues = values
            this.countryCode = this.getValue<string>(rawValues, BomTelephoneNumber.COUNTRY_CODE_FIELD_NAME);
            this.extension = this.getValue<string>(rawValues, BomTelephoneNumber.EXTENSION_FIELD_NAME);
            this.number = this.getValue<string>(rawValues, BomTelephoneNumber.NUMBER_FIELD_NAME);
            this.type = this.getValue<TelecomType>(rawValues, BomTelephoneNumber.TYPE_FIELD_NAME);
        }
    }
}
