// swagger-ts-generator/test-gen/app/models/webapi/bom-change-notification.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';

import { IBomChangeNotificationEntity, BomChangeNotificationEntity } from './bom-change-notification-entity.model';

export interface IBomChangeNotification {
    commandType?: string;
    connectionId?: string;
    entity?: IBomChangeNotificationEntity;
    nextActionExpectedFromMe?: boolean;
    note?: string;
    userNickName?: string;
}


export class BomChangeNotification extends BaseModel implements IBomChangeNotification  {

    static COMMAND_TYPE_FIELD_NAME = 'commandType';
    static CONNECTION_ID_FIELD_NAME = 'connectionId';
    static ENTITY_FIELD_NAME = 'entity';
    static NEXT_ACTION_EXPECTED_FROM_ME_FIELD_NAME = 'nextActionExpectedFromMe';
    static NOTE_FIELD_NAME = 'note';
    static USER_NICK_NAME_FIELD_NAME = 'userNickName';

    commandType: string;
    connectionId: string;
    entity: BomChangeNotificationEntity;
    nextActionExpectedFromMe: boolean;
    note: string;
    userNickName: string;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    */
    constructor(values?: Partial<IBomChangeNotification>) {
        super();
        this.entity = new BomChangeNotificationEntity(); 

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<IBomChangeNotification>): void {
        if (values) {
            
            const rawValues = values
            this.commandType = this.getValue<string>(rawValues, BomChangeNotification.COMMAND_TYPE_FIELD_NAME);
            this.connectionId = this.getValue<string>(rawValues, BomChangeNotification.CONNECTION_ID_FIELD_NAME);
            this.entity.setValues(rawValues.entity);
            this.nextActionExpectedFromMe = this.getValue<boolean>(rawValues, BomChangeNotification.NEXT_ACTION_EXPECTED_FROM_ME_FIELD_NAME);
            this.note = this.getValue<string>(rawValues, BomChangeNotification.NOTE_FIELD_NAME);
            this.userNickName = this.getValue<string>(rawValues, BomChangeNotification.USER_NICK_NAME_FIELD_NAME);
        }
    }
}
