// swagger-ts-generator/test-gen/app/models/webapi/bom-reinsurance-treaty.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';

import { IBomParticipantInRole, BomParticipantInRole } from './bom-participant-in-role.model';
import { IBomDocument, BomDocument } from './bom-document.model';
import { IBomBoundedPeriod, BomBoundedPeriod } from './bom-bounded-period.model';
import { IBomReinsurancePortfolio, BomReinsurancePortfolio } from './bom-reinsurance-portfolio.model';

export interface IBomReinsuranceTreaty {
    cedent?: IBomParticipantInRole;
    createdAt?: Date;
    description?: string;
    documents?: Array<IBomDocument>;
    effectivePeriod?: IBomBoundedPeriod;
    id?: string;
    lastModifiedAt?: Date;
    name: string;
    portfolios?: Array<IBomReinsurancePortfolio>;
    referenceCurrency?: string;
}


export class BomReinsuranceTreaty extends BaseModel implements IBomReinsuranceTreaty  {

    static CEDENT_FIELD_NAME = 'cedent';
    static CREATED_AT_FIELD_NAME = 'createdAt';
    static DESCRIPTION_FIELD_NAME = 'description';
    static DOCUMENTS_FIELD_NAME = 'documents';
    static EFFECTIVE_PERIOD_FIELD_NAME = 'effectivePeriod';
    static ID_FIELD_NAME = 'id';
    static LAST_MODIFIED_AT_FIELD_NAME = 'lastModifiedAt';
    static NAME_FIELD_NAME = 'name';
    static PORTFOLIOS_FIELD_NAME = 'portfolios';
    static REFERENCE_CURRENCY_FIELD_NAME = 'referenceCurrency';

    cedent: BomParticipantInRole;
    createdAt: Date;
    description: string;
    documents: Array<BomDocument>;
    effectivePeriod: BomBoundedPeriod;
    id: string;
    lastModifiedAt: Date;
    name: string;
    portfolios: Array<BomReinsurancePortfolio>;
    referenceCurrency: string;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    */
    constructor(values?: Partial<IBomReinsuranceTreaty>) {
        super();
        this.cedent = new BomParticipantInRole(); 
        this.documents = new Array<BomDocument>(); 
        this.effectivePeriod = new BomBoundedPeriod(); 
        this.portfolios = new Array<BomReinsurancePortfolio>(); 

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<IBomReinsuranceTreaty>): void {
        if (values) {
            
            const rawValues = values
            this.cedent.setValues(rawValues.cedent);
            this.createdAt = this.getValue<Date>(rawValues, BomReinsuranceTreaty.CREATED_AT_FIELD_NAME);
            this.description = this.getValue<string>(rawValues, BomReinsuranceTreaty.DESCRIPTION_FIELD_NAME);
            this.fillModelArray<IBomDocument>(this, BomReinsuranceTreaty.DOCUMENTS_FIELD_NAME, rawValues.documents, BomDocument, SubTypeFactory.createSubTypeInstance);
            this.effectivePeriod.setValues(rawValues.effectivePeriod);
            this.id = this.getValue<string>(rawValues, BomReinsuranceTreaty.ID_FIELD_NAME);
            this.lastModifiedAt = this.getValue<Date>(rawValues, BomReinsuranceTreaty.LAST_MODIFIED_AT_FIELD_NAME);
            this.name = this.getValue<string>(rawValues, BomReinsuranceTreaty.NAME_FIELD_NAME);
            this.fillModelArray<IBomReinsurancePortfolio>(this, BomReinsuranceTreaty.PORTFOLIOS_FIELD_NAME, rawValues.portfolios, BomReinsurancePortfolio, SubTypeFactory.createSubTypeInstance);
            this.referenceCurrency = this.getValue<string>(rawValues, BomReinsuranceTreaty.REFERENCE_CURRENCY_FIELD_NAME);
        }
    }
}
