// swagger-ts-generator/test-gen/app/models/webapi/bom-reinsurance-insurance-service-request.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';

import { IBomDocument, BomDocument } from './bom-document.model';
import { IBomParticipantInRole, BomParticipantInRole } from './bom-participant-in-role.model';
import { InsuranceServiceRequestStatus } from './enums';
import { BomTransactionStatus } from './enums';
import { InsuranceServiceRequestType } from './enums';

export interface IBomReinsuranceInsuranceServiceRequest {
    createdAt?: Date;
    description?: string;
    documents?: Array<IBomDocument>;
    followUpDueAt?: Date;
    id?: string;
    lastModifiedAt?: Date;
    myCompanyParticipantInRole?: IBomParticipantInRole;
    participants?: Array<IBomParticipantInRole>;
    status?: InsuranceServiceRequestStatus;
    transactionStatus?: BomTransactionStatus;
    treatyId?: string;
    type?: InsuranceServiceRequestType;
}


export class BomReinsuranceInsuranceServiceRequest extends BaseModel implements IBomReinsuranceInsuranceServiceRequest  {

    static CREATED_AT_FIELD_NAME = 'createdAt';
    static DESCRIPTION_FIELD_NAME = 'description';
    static DOCUMENTS_FIELD_NAME = 'documents';
    static FOLLOW_UP_DUE_AT_FIELD_NAME = 'followUpDueAt';
    static ID_FIELD_NAME = 'id';
    static LAST_MODIFIED_AT_FIELD_NAME = 'lastModifiedAt';
    static MY_COMPANY_PARTICIPANT_IN_ROLE_FIELD_NAME = 'myCompanyParticipantInRole';
    static PARTICIPANTS_FIELD_NAME = 'participants';
    static STATUS_FIELD_NAME = 'status';
    static TRANSACTION_STATUS_FIELD_NAME = 'transactionStatus';
    static TREATY_ID_FIELD_NAME = 'treatyId';
    static TYPE_FIELD_NAME = 'type';

    createdAt: Date;
    description: string;
    documents: Array<BomDocument>;
    followUpDueAt: Date;
    id: string;
    lastModifiedAt: Date;
    myCompanyParticipantInRole: BomParticipantInRole;
    participants: Array<BomParticipantInRole>;
    status: InsuranceServiceRequestStatus;
    transactionStatus: BomTransactionStatus;
    treatyId: string;
    type: InsuranceServiceRequestType;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    */
    constructor(values?: Partial<IBomReinsuranceInsuranceServiceRequest>) {
        super();
        this.documents = new Array<BomDocument>(); 
        this.myCompanyParticipantInRole = new BomParticipantInRole(); 
        this.participants = new Array<BomParticipantInRole>(); 

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<IBomReinsuranceInsuranceServiceRequest>): void {
        if (values) {
            
            const rawValues = values
            this.createdAt = this.getValue<Date>(rawValues, BomReinsuranceInsuranceServiceRequest.CREATED_AT_FIELD_NAME);
            this.description = this.getValue<string>(rawValues, BomReinsuranceInsuranceServiceRequest.DESCRIPTION_FIELD_NAME);
            this.fillModelArray<IBomDocument>(this, BomReinsuranceInsuranceServiceRequest.DOCUMENTS_FIELD_NAME, rawValues.documents, BomDocument, SubTypeFactory.createSubTypeInstance);
            this.followUpDueAt = this.getValue<Date>(rawValues, BomReinsuranceInsuranceServiceRequest.FOLLOW_UP_DUE_AT_FIELD_NAME);
            this.id = this.getValue<string>(rawValues, BomReinsuranceInsuranceServiceRequest.ID_FIELD_NAME);
            this.lastModifiedAt = this.getValue<Date>(rawValues, BomReinsuranceInsuranceServiceRequest.LAST_MODIFIED_AT_FIELD_NAME);
            this.myCompanyParticipantInRole.setValues(rawValues.myCompanyParticipantInRole);
            this.fillModelArray<IBomParticipantInRole>(this, BomReinsuranceInsuranceServiceRequest.PARTICIPANTS_FIELD_NAME, rawValues.participants, BomParticipantInRole, SubTypeFactory.createSubTypeInstance);
            this.status = this.getValue<InsuranceServiceRequestStatus>(rawValues, BomReinsuranceInsuranceServiceRequest.STATUS_FIELD_NAME);
            this.transactionStatus = this.getValue<BomTransactionStatus>(rawValues, BomReinsuranceInsuranceServiceRequest.TRANSACTION_STATUS_FIELD_NAME);
            this.treatyId = this.getValue<string>(rawValues, BomReinsuranceInsuranceServiceRequest.TREATY_ID_FIELD_NAME);
            this.type = this.getValue<InsuranceServiceRequestType>(rawValues, BomReinsuranceInsuranceServiceRequest.TYPE_FIELD_NAME);
        }
    }
}
