// swagger-ts-generator/test-gen/app/models/webapi/bom-change-notification-entity.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';

import { Codex1EntityTypeSelector } from './enums';

export interface IBomChangeNotificationEntity {
    dateTime?: Date;
    description?: string;
    entityTypeSelector?: Codex1EntityTypeSelector;
    id?: string;
    name?: string;
    parentEntity?: IBomChangeNotificationEntity;
}


export class BomChangeNotificationEntity extends BaseModel implements IBomChangeNotificationEntity  {

    static DATE_TIME_FIELD_NAME = 'dateTime';
    static DESCRIPTION_FIELD_NAME = 'description';
    static ENTITY_TYPE_SELECTOR_FIELD_NAME = 'entityTypeSelector';
    static ID_FIELD_NAME = 'id';
    static NAME_FIELD_NAME = 'name';
    static PARENT_ENTITY_FIELD_NAME = 'parentEntity';

    dateTime: Date;
    description: string;
    entityTypeSelector: Codex1EntityTypeSelector;
    id: string;
    name: string;
    parentEntity: BomChangeNotificationEntity;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    */
    constructor(values?: Partial<IBomChangeNotificationEntity>) {
        super();
        this.parentEntity = new BomChangeNotificationEntity(); 

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<IBomChangeNotificationEntity>): void {
        if (values) {
            
            const rawValues = values
            this.dateTime = this.getValue<Date>(rawValues, BomChangeNotificationEntity.DATE_TIME_FIELD_NAME);
            this.description = this.getValue<string>(rawValues, BomChangeNotificationEntity.DESCRIPTION_FIELD_NAME);
            this.entityTypeSelector = this.getValue<Codex1EntityTypeSelector>(rawValues, BomChangeNotificationEntity.ENTITY_TYPE_SELECTOR_FIELD_NAME);
            this.id = this.getValue<string>(rawValues, BomChangeNotificationEntity.ID_FIELD_NAME);
            this.name = this.getValue<string>(rawValues, BomChangeNotificationEntity.NAME_FIELD_NAME);
            this.parentEntity.setValues(rawValues.parentEntity);
        }
    }
}
