// swagger-ts-generator/test-gen/app/models/webapi/bom-quota-share-cash-loss-entry.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';

import { IMoney, Money } from './money.model';

export interface IBomQuotaShareCashLossEntry {
    cashLossEntryId?: string;
    lossAmount?: IMoney;
    lossRate?: string;
    sections?: Array<string>;
}


export class BomQuotaShareCashLossEntry extends BaseModel implements IBomQuotaShareCashLossEntry  {

    static CASH_LOSS_ENTRY_ID_FIELD_NAME = 'cashLossEntryId';
    static LOSS_AMOUNT_FIELD_NAME = 'lossAmount';
    static LOSS_RATE_FIELD_NAME = 'lossRate';
    static SECTIONS_FIELD_NAME = 'sections';

    cashLossEntryId: string;
    lossAmount: Money;
    lossRate: string;
    sections: Array<string>;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    */
    constructor(values?: Partial<IBomQuotaShareCashLossEntry>) {
        super();
        this.lossAmount = new Money(); 
        this.sections = new Array<string>(); 

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<IBomQuotaShareCashLossEntry>): void {
        if (values) {
            
            const rawValues = values
            this.cashLossEntryId = this.getValue<string>(rawValues, BomQuotaShareCashLossEntry.CASH_LOSS_ENTRY_ID_FIELD_NAME);
            this.lossAmount.setValues(rawValues.lossAmount);
            this.lossRate = this.getValue<string>(rawValues, BomQuotaShareCashLossEntry.LOSS_RATE_FIELD_NAME);
            this.fillModelArray<string>(this, BomQuotaShareCashLossEntry.SECTIONS_FIELD_NAME, rawValues.sections);
        }
    }
}
