// swagger-ts-generator/test-gen/app/models/webapi/bom-reinsurance-treaty-summary.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';

import { IBomParticipantInRole, BomParticipantInRole } from './bom-participant-in-role.model';
import { IBomBoundedPeriod, BomBoundedPeriod } from './bom-bounded-period.model';
import { BomParticipantRole } from './enums';
import { BomReinsuranceTreatyStatus } from './enums';

export interface IBomReinsuranceTreatySummary {
    brokers?: Array<IBomParticipantInRole>;
    cedent?: IBomParticipantInRole;
    effectivePeriod?: IBomBoundedPeriod;
    id?: string;
    lastModifiedAt?: Date;
    myCompanyParticipantRole?: BomParticipantRole;
    name?: string;
    nrOfOpenTasks?: number;
    nrOfUnreadMessages?: number;
    reinsurers?: Array<IBomParticipantInRole>;
    status?: BomReinsuranceTreatyStatus;
}


export class BomReinsuranceTreatySummary extends BaseModel implements IBomReinsuranceTreatySummary  {

    static BROKERS_FIELD_NAME = 'brokers';
    static CEDENT_FIELD_NAME = 'cedent';
    static EFFECTIVE_PERIOD_FIELD_NAME = 'effectivePeriod';
    static ID_FIELD_NAME = 'id';
    static LAST_MODIFIED_AT_FIELD_NAME = 'lastModifiedAt';
    static MY_COMPANY_PARTICIPANT_ROLE_FIELD_NAME = 'myCompanyParticipantRole';
    static NAME_FIELD_NAME = 'name';
    static NR_OF_OPEN_TASKS_FIELD_NAME = 'nrOfOpenTasks';
    static NR_OF_UNREAD_MESSAGES_FIELD_NAME = 'nrOfUnreadMessages';
    static REINSURERS_FIELD_NAME = 'reinsurers';
    static STATUS_FIELD_NAME = 'status';

    brokers: Array<BomParticipantInRole>;
    cedent: BomParticipantInRole;
    effectivePeriod: BomBoundedPeriod;
    id: string;
    lastModifiedAt: Date;
    myCompanyParticipantRole: BomParticipantRole;
    name: string;
    nrOfOpenTasks: number;
    nrOfUnreadMessages: number;
    reinsurers: Array<BomParticipantInRole>;
    status: BomReinsuranceTreatyStatus;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    */
    constructor(values?: Partial<IBomReinsuranceTreatySummary>) {
        super();
        this.brokers = new Array<BomParticipantInRole>(); 
        this.cedent = new BomParticipantInRole(); 
        this.effectivePeriod = new BomBoundedPeriod(); 
        this.reinsurers = new Array<BomParticipantInRole>(); 

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<IBomReinsuranceTreatySummary>): void {
        if (values) {
            
            const rawValues = values
            this.fillModelArray<IBomParticipantInRole>(this, BomReinsuranceTreatySummary.BROKERS_FIELD_NAME, rawValues.brokers, BomParticipantInRole, SubTypeFactory.createSubTypeInstance);
            this.cedent.setValues(rawValues.cedent);
            this.effectivePeriod.setValues(rawValues.effectivePeriod);
            this.id = this.getValue<string>(rawValues, BomReinsuranceTreatySummary.ID_FIELD_NAME);
            this.lastModifiedAt = this.getValue<Date>(rawValues, BomReinsuranceTreatySummary.LAST_MODIFIED_AT_FIELD_NAME);
            this.myCompanyParticipantRole = this.getValue<BomParticipantRole>(rawValues, BomReinsuranceTreatySummary.MY_COMPANY_PARTICIPANT_ROLE_FIELD_NAME);
            this.name = this.getValue<string>(rawValues, BomReinsuranceTreatySummary.NAME_FIELD_NAME);
            this.nrOfOpenTasks = this.getValue<number>(rawValues, BomReinsuranceTreatySummary.NR_OF_OPEN_TASKS_FIELD_NAME);
            this.nrOfUnreadMessages = this.getValue<number>(rawValues, BomReinsuranceTreatySummary.NR_OF_UNREAD_MESSAGES_FIELD_NAME);
            this.fillModelArray<IBomParticipantInRole>(this, BomReinsuranceTreatySummary.REINSURERS_FIELD_NAME, rawValues.reinsurers, BomParticipantInRole, SubTypeFactory.createSubTypeInstance);
            this.status = this.getValue<BomReinsuranceTreatyStatus>(rawValues, BomReinsuranceTreatySummary.STATUS_FIELD_NAME);
        }
    }
}
