// swagger-ts-generator/test-gen/app/models/webapi/bom-reinsurance-reinstatement.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';

import { IMoney, Money } from './money.model';

export interface IBomReinsuranceReinstatement {
    reinstatementCapacityAmount?: IMoney;
    reinstatementCapacityPercentage?: string;
    reinstatementPremiumAmount?: IMoney;
    reinstatementPremiumPercentage?: string;
}


export class BomReinsuranceReinstatement extends BaseModel implements IBomReinsuranceReinstatement  {

    static REINSTATEMENT_CAPACITY_AMOUNT_FIELD_NAME = 'reinstatementCapacityAmount';
    static REINSTATEMENT_CAPACITY_PERCENTAGE_FIELD_NAME = 'reinstatementCapacityPercentage';
    static REINSTATEMENT_PREMIUM_AMOUNT_FIELD_NAME = 'reinstatementPremiumAmount';
    static REINSTATEMENT_PREMIUM_PERCENTAGE_FIELD_NAME = 'reinstatementPremiumPercentage';

    reinstatementCapacityAmount: Money;
    reinstatementCapacityPercentage: string;
    reinstatementPremiumAmount: Money;
    reinstatementPremiumPercentage: string;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    */
    constructor(values?: Partial<IBomReinsuranceReinstatement>) {
        super();
        this.reinstatementCapacityAmount = new Money(); 
        this.reinstatementPremiumAmount = new Money(); 

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<IBomReinsuranceReinstatement>): void {
        if (values) {
            
            const rawValues = values
            this.reinstatementCapacityAmount.setValues(rawValues.reinstatementCapacityAmount);
            this.reinstatementCapacityPercentage = this.getValue<string>(rawValues, BomReinsuranceReinstatement.REINSTATEMENT_CAPACITY_PERCENTAGE_FIELD_NAME);
            this.reinstatementPremiumAmount.setValues(rawValues.reinstatementPremiumAmount);
            this.reinstatementPremiumPercentage = this.getValue<string>(rawValues, BomReinsuranceReinstatement.REINSTATEMENT_PREMIUM_PERCENTAGE_FIELD_NAME);
        }
    }
}
