// swagger-ts-generator/test-gen/app/models/webapi/bom-reinsurance-portfolio.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';

import { IMoney, Money } from './money.model';
import { BomClassOfBusiness } from './enums';

export interface IBomReinsurancePortfolio {
    classesOfBusiness: Array<BomClassOfBusiness>;
    createdAt?: Date;
    description: string;
    estimatedPremiumIncome?: IMoney;
    id?: string;
    lastModifiedAt?: Date;
    name: string;
    regions: Array<string>;
    totalSumInsured?: IMoney;
    version?: number;
}


export class BomReinsurancePortfolio extends BaseModel implements IBomReinsurancePortfolio  {

    static CLASSES_OF_BUSINESS_FIELD_NAME = 'classesOfBusiness';
    static CREATED_AT_FIELD_NAME = 'createdAt';
    static DESCRIPTION_FIELD_NAME = 'description';
    static ESTIMATED_PREMIUM_INCOME_FIELD_NAME = 'estimatedPremiumIncome';
    static ID_FIELD_NAME = 'id';
    static LAST_MODIFIED_AT_FIELD_NAME = 'lastModifiedAt';
    static NAME_FIELD_NAME = 'name';
    static REGIONS_FIELD_NAME = 'regions';
    static TOTAL_SUM_INSURED_FIELD_NAME = 'totalSumInsured';
    static VERSION_FIELD_NAME = 'version';

    classesOfBusiness: Array<BomClassOfBusiness>;
    createdAt: Date;
    description: string;
    estimatedPremiumIncome: Money;
    id: string;
    lastModifiedAt: Date;
    name: string;
    regions: Array<string>;
    totalSumInsured: Money;
    version: number;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    */
    constructor(values?: Partial<IBomReinsurancePortfolio>) {
        super();
        this.classesOfBusiness = new Array<BomClassOfBusiness>(); 
        this.estimatedPremiumIncome = new Money(); 
        this.regions = new Array<string>(); 
        this.totalSumInsured = new Money(); 

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<IBomReinsurancePortfolio>): void {
        if (values) {
            
            const rawValues = values
            this.fillModelArray<BomClassOfBusiness>(this, BomReinsurancePortfolio.CLASSES_OF_BUSINESS_FIELD_NAME, rawValues.classesOfBusiness);
            this.createdAt = this.getValue<Date>(rawValues, BomReinsurancePortfolio.CREATED_AT_FIELD_NAME);
            this.description = this.getValue<string>(rawValues, BomReinsurancePortfolio.DESCRIPTION_FIELD_NAME);
            this.estimatedPremiumIncome.setValues(rawValues.estimatedPremiumIncome);
            this.id = this.getValue<string>(rawValues, BomReinsurancePortfolio.ID_FIELD_NAME);
            this.lastModifiedAt = this.getValue<Date>(rawValues, BomReinsurancePortfolio.LAST_MODIFIED_AT_FIELD_NAME);
            this.name = this.getValue<string>(rawValues, BomReinsurancePortfolio.NAME_FIELD_NAME);
            this.fillModelArray<string>(this, BomReinsurancePortfolio.REGIONS_FIELD_NAME, rawValues.regions);
            this.totalSumInsured.setValues(rawValues.totalSumInsured);
            this.version = this.getValue<number>(rawValues, BomReinsurancePortfolio.VERSION_FIELD_NAME);
        }
    }
}
