// swagger-ts-generator/test-gen/app/models/webapi/bom-quota-share-section.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';

import { IBomBaseReinsuranceSection, BomBaseReinsuranceSection } from './bom-base-reinsurance-section.model';
import { IBomQuotaShareCashLossEntry, BomQuotaShareCashLossEntry } from './bom-quota-share-cash-loss-entry.model';
import { ILimit, Limit } from './limit.model';
import { IBomQuotaShareLossAdviceEntry, BomQuotaShareLossAdviceEntry } from './bom-quota-share-loss-advice-entry.model';
import { IBomCedingCommission, BomCedingCommission } from './bom-ceding-commission.model';
import { BomClassOfBusiness } from './enums';

export interface IBomQuotaShareSection extends IBomBaseReinsuranceSection{
    retentionPercentage?: string;
    cashLosses?: Array<IBomQuotaShareCashLossEntry>;
    cessionPercentage?: string;
    mainLineOfBusiness?: BomClassOfBusiness;
    territorialScopes?: Array<string>;
    limits?: Array<ILimit>;
    lossAdviceEntry?: IBomQuotaShareLossAdviceEntry;
    cedingCommission?: IBomCedingCommission;
}


export class BomQuotaShareSection extends BomBaseReinsuranceSection implements IBomQuotaShareSection  {
    static TYPE_SELECTOR = 'BomQuotaShareSection';

    static RETENTION_PERCENTAGE_FIELD_NAME = 'retentionPercentage';
    static CASH_LOSSES_FIELD_NAME = 'cashLosses';
    static CESSION_PERCENTAGE_FIELD_NAME = 'cessionPercentage';
    static MAIN_LINE_OF_BUSINESS_FIELD_NAME = 'mainLineOfBusiness';
    static TERRITORIAL_SCOPES_FIELD_NAME = 'territorialScopes';
    static LIMITS_FIELD_NAME = 'limits';
    static LOSS_ADVICE_ENTRY_FIELD_NAME = 'lossAdviceEntry';
    static CEDING_COMMISSION_FIELD_NAME = 'cedingCommission';

    retentionPercentage: string;
    cashLosses: Array<BomQuotaShareCashLossEntry>;
    cessionPercentage: string;
    mainLineOfBusiness: BomClassOfBusiness;
    territorialScopes: Array<string>;
    limits: Array<Limit>;
    lossAdviceEntry: BomQuotaShareLossAdviceEntry;
    cedingCommission: BomCedingCommission;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    */
    constructor(values?: Partial<IBomQuotaShareSection>) {
        super();

        this.typeSelector = BomQuotaShareSection.TYPE_SELECTOR;
        this.cashLosses = new Array<BomQuotaShareCashLossEntry>(); 
        this.territorialScopes = new Array<string>(); 
        this.limits = new Array<Limit>(); 
        this.lossAdviceEntry = new BomQuotaShareLossAdviceEntry(); 
        this.cedingCommission = new BomCedingCommission(); 

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<IBomQuotaShareSection>): void {
        if (values) {
            super.setValues(values)
            const rawValues = values
            this.retentionPercentage = this.getValue<string>(rawValues, BomQuotaShareSection.RETENTION_PERCENTAGE_FIELD_NAME);
            this.fillModelArray<IBomQuotaShareCashLossEntry>(this, BomQuotaShareSection.CASH_LOSSES_FIELD_NAME, rawValues.cashLosses, BomQuotaShareCashLossEntry, SubTypeFactory.createSubTypeInstance);
            this.cessionPercentage = this.getValue<string>(rawValues, BomQuotaShareSection.CESSION_PERCENTAGE_FIELD_NAME);
            this.mainLineOfBusiness = this.getValue<BomClassOfBusiness>(rawValues, BomQuotaShareSection.MAIN_LINE_OF_BUSINESS_FIELD_NAME);
            this.fillModelArray<string>(this, BomQuotaShareSection.TERRITORIAL_SCOPES_FIELD_NAME, rawValues.territorialScopes);
            this.fillModelArray<ILimit>(this, BomQuotaShareSection.LIMITS_FIELD_NAME, rawValues.limits, Limit, SubTypeFactory.createSubTypeInstance);
            this.lossAdviceEntry.setValues(rawValues.lossAdviceEntry);
            this.cedingCommission.setValues(rawValues.cedingCommission);
        }
    }
}
