// swagger-ts-generator/test-gen/app/models/webapi/business-network-and-membership.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';

import { IBusinessNetwork, BusinessNetwork } from './business-network.model';
import { IBusinessNetworkMembership, BusinessNetworkMembership } from './business-network-membership.model';
import { IProtocol, Protocol } from './protocol.model';

export interface IBusinessNetworkAndMembership {
    businessNetwork?: IBusinessNetwork;
    networkMembership?: IBusinessNetworkMembership;
    protocol?: IProtocol;
}


export class BusinessNetworkAndMembership extends BaseModel implements IBusinessNetworkAndMembership  {

    static BUSINESS_NETWORK_FIELD_NAME = 'businessNetwork';
    static NETWORK_MEMBERSHIP_FIELD_NAME = 'networkMembership';
    static PROTOCOL_FIELD_NAME = 'protocol';

    businessNetwork: BusinessNetwork;
    networkMembership: BusinessNetworkMembership;
    protocol: Protocol;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    */
    constructor(values?: Partial<IBusinessNetworkAndMembership>) {
        super();
        this.businessNetwork = new BusinessNetwork(); 
        this.networkMembership = new BusinessNetworkMembership(); 
        this.protocol = new Protocol(); 

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<IBusinessNetworkAndMembership>): void {
        if (values) {
            
            const rawValues = values
            this.businessNetwork.setValues(rawValues.businessNetwork);
            this.networkMembership.setValues(rawValues.networkMembership);
            this.protocol.setValues(rawValues.protocol);
        }
    }
}
