// swagger-ts-generator/test-gen/app/models/webapi/business-network.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';

import { ISoftReference, SoftReference } from './soft-reference.model';

export interface IBusinessNetwork {
    bno?: ISoftReference;
    description?: string;
    id?: string;
    members?: Array<ISoftReference>;
    name?: string;
}


export class BusinessNetwork extends BaseModel implements IBusinessNetwork  {

    static BNO_FIELD_NAME = 'bno';
    static DESCRIPTION_FIELD_NAME = 'description';
    static ID_FIELD_NAME = 'id';
    static MEMBERS_FIELD_NAME = 'members';
    static NAME_FIELD_NAME = 'name';

    bno: SoftReference;
    description: string;
    id: string;
    members: Array<SoftReference>;
    name: string;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    */
    constructor(values?: Partial<IBusinessNetwork>) {
        super();
        this.bno = new SoftReference(); 
        this.members = new Array<SoftReference>(); 

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<IBusinessNetwork>): void {
        if (values) {
            
            const rawValues = values
            this.bno.setValues(rawValues.bno);
            this.description = this.getValue<string>(rawValues, BusinessNetwork.DESCRIPTION_FIELD_NAME);
            this.id = this.getValue<string>(rawValues, BusinessNetwork.ID_FIELD_NAME);
            this.fillModelArray<ISoftReference>(this, BusinessNetwork.MEMBERS_FIELD_NAME, rawValues.members, SoftReference, SubTypeFactory.createSubTypeInstance);
            this.name = this.getValue<string>(rawValues, BusinessNetwork.NAME_FIELD_NAME);
        }
    }
}
