// swagger-ts-generator/test-gen/app/models/webapi/contract-diff-wording.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';

import { IAbstractDiffField, AbstractDiffField } from './abstract-diff-field.model';

export interface IContractDiffWording extends IAbstractDiffField{
    fieldName?: string;
    fieldType?: string;
    previousValue?: object;
    currentValue?: object;
}


export class ContractDiffWording extends AbstractDiffField implements IContractDiffWording  {
    static TYPE_SELECTOR = 'ContractDiffWording';

    static FIELD_NAME_FIELD_NAME = 'fieldName';
    static FIELD_TYPE_FIELD_NAME = 'fieldType';
    static PREVIOUS_VALUE_FIELD_NAME = 'previousValue';
    static CURRENT_VALUE_FIELD_NAME = 'currentValue';

    fieldName: string;
    fieldType: string;
    previousValue: object;
    currentValue: object;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    */
    constructor(values?: Partial<IContractDiffWording>) {
        super();

        this.typeSelector = ContractDiffWording.TYPE_SELECTOR;

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<IContractDiffWording>): void {
        if (values) {
            super.setValues(values)
            const rawValues = values
            this.fieldName = this.getValue<string>(rawValues, ContractDiffWording.FIELD_NAME_FIELD_NAME);
            this.fieldType = this.getValue<string>(rawValues, ContractDiffWording.FIELD_TYPE_FIELD_NAME);
            this.previousValue = this.getValue<object>(rawValues, ContractDiffWording.PREVIOUS_VALUE_FIELD_NAME);
            this.currentValue = this.getValue<object>(rawValues, ContractDiffWording.CURRENT_VALUE_FIELD_NAME);
        }
    }
}
