// swagger-ts-generator/test-gen/app/models/webapi/contract-signing-container.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';

import { IReinsuranceContract, ReinsuranceContract } from './reinsurance-contract.model';
import { IInsuranceContractSigning, InsuranceContractSigning } from './insurance-contract-signing.model';
import { InsuranceContractProcessStatus } from './enums';
import { InsuranceContractSigningCommand } from './enums';
import { TransactionStatus } from './enums';

export interface IContractSigningContainer {
    contract?: IReinsuranceContract;
    contractTargetStatus?: InsuranceContractProcessStatus;
    process?: IInsuranceContractSigning;
    processCommand?: InsuranceContractSigningCommand;
    transactionStatus?: TransactionStatus;
    transitoryState?: boolean;
}


export class ContractSigningContainer extends BaseModel implements IContractSigningContainer  {

    static CONTRACT_FIELD_NAME = 'contract';
    static CONTRACT_TARGET_STATUS_FIELD_NAME = 'contractTargetStatus';
    static PROCESS_FIELD_NAME = 'process';
    static PROCESS_COMMAND_FIELD_NAME = 'processCommand';
    static TRANSACTION_STATUS_FIELD_NAME = 'transactionStatus';
    static TRANSITORY_STATE_FIELD_NAME = 'transitoryState';

    contract: ReinsuranceContract;
    contractTargetStatus: InsuranceContractProcessStatus;
    process: InsuranceContractSigning;
    processCommand: InsuranceContractSigningCommand;
    transactionStatus: TransactionStatus;
    transitoryState: boolean;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    */
    constructor(values?: Partial<IContractSigningContainer>) {
        super();
        this.contract = new ReinsuranceContract(); 
        this.process = new InsuranceContractSigning(); 

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<IContractSigningContainer>): void {
        if (values) {
            
            const rawValues = values
            this.contract.setValues(rawValues.contract);
            this.contractTargetStatus = this.getValue<InsuranceContractProcessStatus>(rawValues, ContractSigningContainer.CONTRACT_TARGET_STATUS_FIELD_NAME);
            this.process.setValues(rawValues.process);
            this.processCommand = this.getValue<InsuranceContractSigningCommand>(rawValues, ContractSigningContainer.PROCESS_COMMAND_FIELD_NAME);
            this.transactionStatus = this.getValue<TransactionStatus>(rawValues, ContractSigningContainer.TRANSACTION_STATUS_FIELD_NAME);
            this.transitoryState = this.getValue<boolean>(rawValues, ContractSigningContainer.TRANSITORY_STATE_FIELD_NAME);
        }
    }
}
