// swagger-ts-generator/test-gen/app/models/webapi/contract-binding-container.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';

import { IReinsuranceContract, ReinsuranceContract } from './reinsurance-contract.model';
import { IInsuranceContractBinding, InsuranceContractBinding } from './insurance-contract-binding.model';
import { InsuranceContractProcessStatus } from './enums';
import { InsuranceContractBindingCommand } from './enums';
import { TransactionStatus } from './enums';

export interface IContractBindingContainer {
    contract?: IReinsuranceContract;
    contractTargetStatus?: InsuranceContractProcessStatus;
    process?: IInsuranceContractBinding;
    processCommand?: InsuranceContractBindingCommand;
    transactionStatus?: TransactionStatus;
    transitoryState?: boolean;
}


export class ContractBindingContainer extends BaseModel implements IContractBindingContainer  {

    static CONTRACT_FIELD_NAME = 'contract';
    static CONTRACT_TARGET_STATUS_FIELD_NAME = 'contractTargetStatus';
    static PROCESS_FIELD_NAME = 'process';
    static PROCESS_COMMAND_FIELD_NAME = 'processCommand';
    static TRANSACTION_STATUS_FIELD_NAME = 'transactionStatus';
    static TRANSITORY_STATE_FIELD_NAME = 'transitoryState';

    contract: ReinsuranceContract;
    contractTargetStatus: InsuranceContractProcessStatus;
    process: InsuranceContractBinding;
    processCommand: InsuranceContractBindingCommand;
    transactionStatus: TransactionStatus;
    transitoryState: boolean;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    */
    constructor(values?: Partial<IContractBindingContainer>) {
        super();
        this.contract = new ReinsuranceContract(); 
        this.process = new InsuranceContractBinding(); 

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<IContractBindingContainer>): void {
        if (values) {
            
            const rawValues = values
            this.contract.setValues(rawValues.contract);
            this.contractTargetStatus = this.getValue<InsuranceContractProcessStatus>(rawValues, ContractBindingContainer.CONTRACT_TARGET_STATUS_FIELD_NAME);
            this.process.setValues(rawValues.process);
            this.processCommand = this.getValue<InsuranceContractBindingCommand>(rawValues, ContractBindingContainer.PROCESS_COMMAND_FIELD_NAME);
            this.transactionStatus = this.getValue<TransactionStatus>(rawValues, ContractBindingContainer.TRANSACTION_STATUS_FIELD_NAME);
            this.transitoryState = this.getValue<boolean>(rawValues, ContractBindingContainer.TRANSITORY_STATE_FIELD_NAME);
        }
    }
}
