// swagger-ts-generator/test-gen/app/models/webapi/contract-diff-clause.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';

import { IAbstractDiffField, AbstractDiffField } from './abstract-diff-field.model';

export interface IContractDiffClause extends IAbstractDiffField{
    name?: string;
    currentName?: string;
    description?: string;
    currentDescription?: string;
    layers?: string;
    currentLayers?: string;
    updated?: boolean;
}


export class ContractDiffClause extends AbstractDiffField implements IContractDiffClause  {
    static TYPE_SELECTOR = 'ContractDiffClause';

    static NAME_FIELD_NAME = 'name';
    static CURRENT_NAME_FIELD_NAME = 'currentName';
    static DESCRIPTION_FIELD_NAME = 'description';
    static CURRENT_DESCRIPTION_FIELD_NAME = 'currentDescription';
    static LAYERS_FIELD_NAME = 'layers';
    static CURRENT_LAYERS_FIELD_NAME = 'currentLayers';
    static UPDATED_FIELD_NAME = 'updated';

    name: string;
    currentName: string;
    description: string;
    currentDescription: string;
    layers: string;
    currentLayers: string;
    updated: boolean;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    */
    constructor(values?: Partial<IContractDiffClause>) {
        super();

        this.typeSelector = ContractDiffClause.TYPE_SELECTOR;

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<IContractDiffClause>): void {
        if (values) {
            super.setValues(values)
            const rawValues = values
            this.name = this.getValue<string>(rawValues, ContractDiffClause.NAME_FIELD_NAME);
            this.currentName = this.getValue<string>(rawValues, ContractDiffClause.CURRENT_NAME_FIELD_NAME);
            this.description = this.getValue<string>(rawValues, ContractDiffClause.DESCRIPTION_FIELD_NAME);
            this.currentDescription = this.getValue<string>(rawValues, ContractDiffClause.CURRENT_DESCRIPTION_FIELD_NAME);
            this.layers = this.getValue<string>(rawValues, ContractDiffClause.LAYERS_FIELD_NAME);
            this.currentLayers = this.getValue<string>(rawValues, ContractDiffClause.CURRENT_LAYERS_FIELD_NAME);
            this.updated = this.getValue<boolean>(rawValues, ContractDiffClause.UPDATED_FIELD_NAME);
        }
    }
}
