// swagger-ts-generator/test-gen/app/models/webapi/contract-quotation-container.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';

import { IReinsuranceContract, ReinsuranceContract } from './reinsurance-contract.model';
import { IInsuranceContractQuotation, InsuranceContractQuotation } from './insurance-contract-quotation.model';
import { InsuranceContractProcessStatus } from './enums';
import { InsuranceContractProcessCommand } from './enums';
import { TransactionStatus } from './enums';

export interface IContractQuotationContainer {
    contract?: IReinsuranceContract;
    contractTargetStatus?: InsuranceContractProcessStatus;
    process?: IInsuranceContractQuotation;
    processCommand?: InsuranceContractProcessCommand;
    transactionStatus?: TransactionStatus;
    transitoryState?: boolean;
}


export class ContractQuotationContainer extends BaseModel implements IContractQuotationContainer  {

    static CONTRACT_FIELD_NAME = 'contract';
    static CONTRACT_TARGET_STATUS_FIELD_NAME = 'contractTargetStatus';
    static PROCESS_FIELD_NAME = 'process';
    static PROCESS_COMMAND_FIELD_NAME = 'processCommand';
    static TRANSACTION_STATUS_FIELD_NAME = 'transactionStatus';
    static TRANSITORY_STATE_FIELD_NAME = 'transitoryState';

    contract: ReinsuranceContract;
    contractTargetStatus: InsuranceContractProcessStatus;
    process: InsuranceContractQuotation;
    processCommand: InsuranceContractProcessCommand;
    transactionStatus: TransactionStatus;
    transitoryState: boolean;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    */
    constructor(values?: Partial<IContractQuotationContainer>) {
        super();
        this.contract = new ReinsuranceContract(); 
        this.process = new InsuranceContractQuotation(); 

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<IContractQuotationContainer>): void {
        if (values) {
            
            const rawValues = values
            this.contract.setValues(rawValues.contract);
            this.contractTargetStatus = this.getValue<InsuranceContractProcessStatus>(rawValues, ContractQuotationContainer.CONTRACT_TARGET_STATUS_FIELD_NAME);
            this.process.setValues(rawValues.process);
            this.processCommand = this.getValue<InsuranceContractProcessCommand>(rawValues, ContractQuotationContainer.PROCESS_COMMAND_FIELD_NAME);
            this.transactionStatus = this.getValue<TransactionStatus>(rawValues, ContractQuotationContainer.TRANSACTION_STATUS_FIELD_NAME);
            this.transitoryState = this.getValue<boolean>(rawValues, ContractQuotationContainer.TRANSITORY_STATE_FIELD_NAME);
        }
    }
}
