// swagger-ts-generator/test-gen/app/models/webapi/corda-tx-execution-state-web-socket-notification.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';

import { WebSocketEventType } from './enums';

export interface ICordaTxExecutionStateWebSocketNotification {
    body?: string;
    progress?: string;
    stateId?: string;
    tag?: string;
    webSocketEventType?: WebSocketEventType;
}


export class CordaTxExecutionStateWebSocketNotification extends BaseModel implements ICordaTxExecutionStateWebSocketNotification  {

    static BODY_FIELD_NAME = 'body';
    static PROGRESS_FIELD_NAME = 'progress';
    static STATE_ID_FIELD_NAME = 'stateId';
    static TAG_FIELD_NAME = 'tag';
    static WEB_SOCKET_EVENT_TYPE_FIELD_NAME = 'webSocketEventType';

    body: string;
    progress: string;
    stateId: string;
    tag: string;
    webSocketEventType: WebSocketEventType;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    */
    constructor(values?: Partial<ICordaTxExecutionStateWebSocketNotification>) {
        super();

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<ICordaTxExecutionStateWebSocketNotification>): void {
        if (values) {
            
            const rawValues = values
            this.body = this.getValue<string>(rawValues, CordaTxExecutionStateWebSocketNotification.BODY_FIELD_NAME);
            this.progress = this.getValue<string>(rawValues, CordaTxExecutionStateWebSocketNotification.PROGRESS_FIELD_NAME);
            this.stateId = this.getValue<string>(rawValues, CordaTxExecutionStateWebSocketNotification.STATE_ID_FIELD_NAME);
            this.tag = this.getValue<string>(rawValues, CordaTxExecutionStateWebSocketNotification.TAG_FIELD_NAME);
            this.webSocketEventType = this.getValue<WebSocketEventType>(rawValues, CordaTxExecutionStateWebSocketNotification.WEB_SOCKET_EVENT_TYPE_FIELD_NAME);
        }
    }
}
