// swagger-ts-generator/test-gen/app/models/webapi/country-subdivision.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';


export interface ICountrySubdivision {
    code?: string;
    name?: string;
    typeSelector?: string;
}


export class CountrySubdivision extends BaseModel implements ICountrySubdivision  {
    static TYPE_SELECTOR = 'CountrySubdivision';

    static CODE_FIELD_NAME = 'code';
    static NAME_FIELD_NAME = 'name';
    static TYPE_SELECTOR_FIELD_NAME = 'typeSelector';

    code: string;
    name: string;
    typeSelector: string;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    */
    constructor(values?: Partial<ICountrySubdivision>) {
        super();

        this.typeSelector = CountrySubdivision.TYPE_SELECTOR;

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<ICountrySubdivision>): void {
        if (values) {
            
            const rawValues = values
            this.code = this.getValue<string>(rawValues, CountrySubdivision.CODE_FIELD_NAME);
            this.name = this.getValue<string>(rawValues, CountrySubdivision.NAME_FIELD_NAME);
            this.typeSelector = this.getValue<string>(rawValues, CountrySubdivision.TYPE_SELECTOR_FIELD_NAME);
        }
    }
}
