// swagger-ts-generator/test-gen/app/models/webapi/enums.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */

// generate enum based on strings instead of numbers
// (see https://blog.rsuter.com/how-to-implement-an-enum-with-string-values-in-typescript/)
export enum AccountCommand {
    ACCEPT_ACCOUNT_ITEMS = <any>"ACCEPT_ACCOUNT_ITEMS",
    QUERY_ACCOUNT_ITEM = <any>"QUERY_ACCOUNT_ITEM",
    RESPOND_TO_QUERY_ACCOUNT_ITEM = <any>"RESPOND_TO_QUERY_ACCOUNT_ITEM",
}

export enum AccountItemCategory {
    CASH = <any>"CASH",
    RESERVE = <any>"RESERVE",
}

export enum AccountItemDueStatus {
    DUE = <any>"DUE",
    NOT_YET_DUE = <any>"NOT_YET_DUE",
}

export enum AccountItemSortProperty {
    ACCOUNTING_PERIOD = <any>"ACCOUNTING_PERIOD",
    ACCOUNT_ID = <any>"ACCOUNT_ID",
    ACCOUNT_ITEM_CATEGORY = <any>"ACCOUNT_ITEM_CATEGORY",
    ACCOUNT_ITEM_ID = <any>"ACCOUNT_ITEM_ID",
    ACCOUNT_ITEM_STATUS = <any>"ACCOUNT_ITEM_STATUS",
    ACCOUNT_ITEM_TYPE = <any>"ACCOUNT_ITEM_TYPE",
    AMOUNT = <any>"AMOUNT",
    CONTRACT_ID = <any>"CONTRACT_ID",
    CONTRACT_NAME = <any>"CONTRACT_NAME",
    COVERAGE_ID = <any>"COVERAGE_ID",
    COVERAGE_NAME = <any>"COVERAGE_NAME",
    EVENT_PROCESSED_DATE = <any>"EVENT_PROCESSED_DATE",
    EVENT_TYPE = <any>"EVENT_TYPE",
    FINANCIAL_EVENT_ID = <any>"FINANCIAL_EVENT_ID",
    PROGRAM_NAME = <any>"PROGRAM_NAME",
}

export enum AccountItemStatus {
    ACCEPTED = <any>"ACCEPTED",
    PENDING = <any>"PENDING",
    QUERIED = <any>"QUERIED",
}

export enum AccountItemType {
    BROKERAGE = <any>"BROKERAGE",
    COMMISSION = <any>"COMMISSION",
    EXPENSE = <any>"EXPENSE",
    LOSS = <any>"LOSS",
    PREMIUM = <any>"PREMIUM",
    REINSTATEMENT = <any>"REINSTATEMENT",
}

export enum AuditDiffEntryType {
    ADDED = <any>"ADDED",
    REMOVED = <any>"REMOVED",
    UPDATED = <any>"UPDATED",
}

export enum BomApplicableLawCode {
    AFGHANISTAN_AF = <any>"AFGHANISTAN_AF",
    ALAND_ISLANDS_AX = <any>"ALAND_ISLANDS_AX",
    ALBANIA_AL = <any>"ALBANIA_AL",
    ALGERIA_DZ = <any>"ALGERIA_DZ",
    AMERICAN_SAMOA_AS = <any>"AMERICAN_SAMOA_AS",
    ANDORRA_AD = <any>"ANDORRA_AD",
    ANGOLA_AO = <any>"ANGOLA_AO",
    ANGUILLA_AI = <any>"ANGUILLA_AI",
    ANTARCTICA_AQ = <any>"ANTARCTICA_AQ",
    ANTIGUA_AND_BARBUDA_AG = <any>"ANTIGUA_AND_BARBUDA_AG",
    ARGENTINA_AR = <any>"ARGENTINA_AR",
    ARMENIA_AM = <any>"ARMENIA_AM",
    ARUBA_AW = <any>"ARUBA_AW",
    AUSTRALIA_AU = <any>"AUSTRALIA_AU",
    AUSTRIA_AT = <any>"AUSTRIA_AT",
    AZERBAIJAN_AZ = <any>"AZERBAIJAN_AZ",
    BAHAMAS_BS = <any>"BAHAMAS_BS",
    BAHRAIN_BH = <any>"BAHRAIN_BH",
    BANGLADESH_BD = <any>"BANGLADESH_BD",
    BARBADOS_BB = <any>"BARBADOS_BB",
    BELARUS_BY = <any>"BELARUS_BY",
    BELGIUM_BE = <any>"BELGIUM_BE",
    BELIZE_BZ = <any>"BELIZE_BZ",
    BENIN_BJ = <any>"BENIN_BJ",
    BERMUDA_BM = <any>"BERMUDA_BM",
    BHUTAN_BT = <any>"BHUTAN_BT",
    BOLIVIA_BO = <any>"BOLIVIA_BO",
    BONAIRE_SINT_EUSTATIUS_AND_SABA_BQ = <any>"BONAIRE_SINT_EUSTATIUS_AND_SABA_BQ",
    BOSNIA_AND_HERZEGOVINA_BA = <any>"BOSNIA_AND_HERZEGOVINA_BA",
    BOTSWANA_BW = <any>"BOTSWANA_BW",
    BOUVET_ISLAND_BV = <any>"BOUVET_ISLAND_BV",
    BRAZIL_BR = <any>"BRAZIL_BR",
    BRITISH_INDIAN_OCEAN_TERRITORY_IO = <any>"BRITISH_INDIAN_OCEAN_TERRITORY_IO",
    BRUNEI_DARUSSALAM_BN = <any>"BRUNEI_DARUSSALAM_BN",
    BULGARIA_BG = <any>"BULGARIA_BG",
    BURKINA_FASO_BF = <any>"BURKINA_FASO_BF",
    BURUNDI_BI = <any>"BURUNDI_BI",
    CABO_VERDE_CV = <any>"CABO_VERDE_CV",
    CAMBODIA_KH = <any>"CAMBODIA_KH",
    CAMEROON_CM = <any>"CAMEROON_CM",
    CANADA_CA = <any>"CANADA_CA",
    CANADA_CA_AB = <any>"CANADA_CA_AB",
    CANADA_CA_BC = <any>"CANADA_CA_BC",
    CANADA_CA_MB = <any>"CANADA_CA_MB",
    CANADA_CA_NB = <any>"CANADA_CA_NB",
    CANADA_CA_NL = <any>"CANADA_CA_NL",
    CANADA_CA_NS = <any>"CANADA_CA_NS",
    CANADA_CA_NT = <any>"CANADA_CA_NT",
    CANADA_CA_NU = <any>"CANADA_CA_NU",
    CANADA_CA_ON = <any>"CANADA_CA_ON",
    CANADA_CA_PE = <any>"CANADA_CA_PE",
    CANADA_CA_QC = <any>"CANADA_CA_QC",
    CANADA_CA_SK = <any>"CANADA_CA_SK",
    CANADA_CA_YT = <any>"CANADA_CA_YT",
    CAYMAN_ISLANDS_KY = <any>"CAYMAN_ISLANDS_KY",
    CENTRAL_AFRICAN_REPUBLIC_CF = <any>"CENTRAL_AFRICAN_REPUBLIC_CF",
    CHAD_TD = <any>"CHAD_TD",
    CHILE_CL = <any>"CHILE_CL",
    CHINA_CN = <any>"CHINA_CN",
    CHRISTMAS_ISLAND_CX = <any>"CHRISTMAS_ISLAND_CX",
    COCOS_ISLANDS_CC = <any>"COCOS_ISLANDS_CC",
    COLOMBIA_CO = <any>"COLOMBIA_CO",
    COMOROS_KM = <any>"COMOROS_KM",
    CONGO_CD = <any>"CONGO_CD",
    CONGO_CG = <any>"CONGO_CG",
    COOK_ISLANDS_CK = <any>"COOK_ISLANDS_CK",
    COSTA_RICA_CR = <any>"COSTA_RICA_CR",
    COTE_D_IVOIRE_CI = <any>"COTE_D_IVOIRE_CI",
    CROATIA_HR = <any>"CROATIA_HR",
    CUBA_CU = <any>"CUBA_CU",
    CURACAO_CW = <any>"CURACAO_CW",
    CYPRUS_CY = <any>"CYPRUS_CY",
    CZECHIA_CZ = <any>"CZECHIA_CZ",
    DENMARK_DK = <any>"DENMARK_DK",
    DJIBOUTI_DJ = <any>"DJIBOUTI_DJ",
    DOMINICAN_REPUBLIC_DO = <any>"DOMINICAN_REPUBLIC_DO",
    DOMINICA_DM = <any>"DOMINICA_DM",
    ECUADOR_EC = <any>"ECUADOR_EC",
    EGYPT_EG = <any>"EGYPT_EG",
    EL_SALVADOR_SV = <any>"EL_SALVADOR_SV",
    ENGLAND_AND_WALES_GB = <any>"ENGLAND_AND_WALES_GB",
    EQUATORIAL_GUINEA_GQ = <any>"EQUATORIAL_GUINEA_GQ",
    ERITREA_ER = <any>"ERITREA_ER",
    ESTONIA_EE = <any>"ESTONIA_EE",
    ESWATINI_SZ = <any>"ESWATINI_SZ",
    ETHIOPIA_ET = <any>"ETHIOPIA_ET",
    FALKLAND_ISLANDS_FK = <any>"FALKLAND_ISLANDS_FK",
    FAROE_ISLANDS_FO = <any>"FAROE_ISLANDS_FO",
    FIJI_FJ = <any>"FIJI_FJ",
    FINLAND_FI = <any>"FINLAND_FI",
    FRANCE_FR = <any>"FRANCE_FR",
    FRENCH_GUIANA_GF = <any>"FRENCH_GUIANA_GF",
    FRENCH_POLYNESIA_PF = <any>"FRENCH_POLYNESIA_PF",
    FRENCH_SOUTHERN_TERRITORIES_TF = <any>"FRENCH_SOUTHERN_TERRITORIES_TF",
    GABON_GA = <any>"GABON_GA",
    GAMBIA_GM = <any>"GAMBIA_GM",
    GEORGIA_GE = <any>"GEORGIA_GE",
    GERMANY_DE = <any>"GERMANY_DE",
    GHANA_GH = <any>"GHANA_GH",
    GIBRALTAR_GI = <any>"GIBRALTAR_GI",
    GREECE_GR = <any>"GREECE_GR",
    GREENLAND_GL = <any>"GREENLAND_GL",
    GRENADA_GD = <any>"GRENADA_GD",
    GUADELOUPE_GP = <any>"GUADELOUPE_GP",
    GUAM_GU = <any>"GUAM_GU",
    GUATEMALA_GT = <any>"GUATEMALA_GT",
    GUERNSEY_GG = <any>"GUERNSEY_GG",
    GUINEA_BISSAU_GW = <any>"GUINEA_BISSAU_GW",
    GUINEA_GN = <any>"GUINEA_GN",
    GUYANA_GY = <any>"GUYANA_GY",
    HAITI_HT = <any>"HAITI_HT",
    HEARD_ISLAND_AND_MCDONALD_ISLANDS_HM = <any>"HEARD_ISLAND_AND_MCDONALD_ISLANDS_HM",
    HOLY_SEE_VA = <any>"HOLY_SEE_VA",
    HONDURAS_HN = <any>"HONDURAS_HN",
    HONG_KONG_HK = <any>"HONG_KONG_HK",
    HUNGARY_HU = <any>"HUNGARY_HU",
    ICELAND_IS = <any>"ICELAND_IS",
    INDIA_IN = <any>"INDIA_IN",
    INDONESIA_ID = <any>"INDONESIA_ID",
    IRAN_IR = <any>"IRAN_IR",
    IRAQ_IQ = <any>"IRAQ_IQ",
    IRELAND_IE = <any>"IRELAND_IE",
    ISLE_OF_MAN_IM = <any>"ISLE_OF_MAN_IM",
    ISRAEL_IL = <any>"ISRAEL_IL",
    ITALY_IT = <any>"ITALY_IT",
    JAMAICA_JM = <any>"JAMAICA_JM",
    JAPAN_JP = <any>"JAPAN_JP",
    JERSEY_JE = <any>"JERSEY_JE",
    JORDAN_JO = <any>"JORDAN_JO",
    KAZAKHSTAN_KZ = <any>"KAZAKHSTAN_KZ",
    KENYA_KE = <any>"KENYA_KE",
    KIRIBATI_KI = <any>"KIRIBATI_KI",
    KOREA_KP = <any>"KOREA_KP",
    KOREA_KR = <any>"KOREA_KR",
    KOSOVO_XK = <any>"KOSOVO_XK",
    KUWAIT_KW = <any>"KUWAIT_KW",
    KYRGYZSTAN_KG = <any>"KYRGYZSTAN_KG",
    LAO_PEOPLES_DEMOCRATIC_REPUBLIC_LA = <any>"LAO_PEOPLES_DEMOCRATIC_REPUBLIC_LA",
    LATVIA_LV = <any>"LATVIA_LV",
    LEBANON_LB = <any>"LEBANON_LB",
    LESOTHO_LS = <any>"LESOTHO_LS",
    LIBERIA_LR = <any>"LIBERIA_LR",
    LIBYA_LY = <any>"LIBYA_LY",
    LIECHTENSTEIN_LI = <any>"LIECHTENSTEIN_LI",
    LITHUANIA_LT = <any>"LITHUANIA_LT",
    LUXEMBOURG_LU = <any>"LUXEMBOURG_LU",
    MACAO_MO = <any>"MACAO_MO",
    MADAGASCAR_MG = <any>"MADAGASCAR_MG",
    MALAWI_MW = <any>"MALAWI_MW",
    MALAYSIA_MY = <any>"MALAYSIA_MY",
    MALDIVES_MV = <any>"MALDIVES_MV",
    MALI_ML = <any>"MALI_ML",
    MALTA_MT = <any>"MALTA_MT",
    MARSHALL_ISLANDS_MH = <any>"MARSHALL_ISLANDS_MH",
    MARTINIQUE_MQ = <any>"MARTINIQUE_MQ",
    MAURITANIA_MR = <any>"MAURITANIA_MR",
    MAURITIUS_MU = <any>"MAURITIUS_MU",
    MAYOTTE_YT = <any>"MAYOTTE_YT",
    MEXICO_MX = <any>"MEXICO_MX",
    MICRONESIA_FM = <any>"MICRONESIA_FM",
    MOLDOVA_MD = <any>"MOLDOVA_MD",
    MONACO_MC = <any>"MONACO_MC",
    MONGOLIA_MN = <any>"MONGOLIA_MN",
    MONTENEGRO_ME = <any>"MONTENEGRO_ME",
    MONTSERRAT_MS = <any>"MONTSERRAT_MS",
    MOROCCO_MA = <any>"MOROCCO_MA",
    MOZAMBIQUE_MZ = <any>"MOZAMBIQUE_MZ",
    MYANMAR_MM = <any>"MYANMAR_MM",
    NAMIBIA_NA = <any>"NAMIBIA_NA",
    NAURU_NR = <any>"NAURU_NR",
    NEPAL_NP = <any>"NEPAL_NP",
    NETHERLANDS_NL = <any>"NETHERLANDS_NL",
    NEW_CALEDONIA_NC = <any>"NEW_CALEDONIA_NC",
    NEW_ZEALAND_NZ = <any>"NEW_ZEALAND_NZ",
    NICARAGUA_NI = <any>"NICARAGUA_NI",
    NIGERIA_NG = <any>"NIGERIA_NG",
    NIGER_NE = <any>"NIGER_NE",
    NIUE_NU = <any>"NIUE_NU",
    NORFOLK_ISLAND_NF = <any>"NORFOLK_ISLAND_NF",
    NORTHERN_MARIANA_ISLANDS_MP = <any>"NORTHERN_MARIANA_ISLANDS_MP",
    NORTH_MACEDONIA_MK = <any>"NORTH_MACEDONIA_MK",
    NORWAY_NO = <any>"NORWAY_NO",
    OMAN_OM = <any>"OMAN_OM",
    PAKISTAN_PK = <any>"PAKISTAN_PK",
    PALAU_PW = <any>"PALAU_PW",
    PALESTINE_PS = <any>"PALESTINE_PS",
    PANAMA_PA = <any>"PANAMA_PA",
    PAPUA_NEW_GUINEA_PG = <any>"PAPUA_NEW_GUINEA_PG",
    PARAGUAY_PY = <any>"PARAGUAY_PY",
    PERU_PE = <any>"PERU_PE",
    PHILIPPINES_PH = <any>"PHILIPPINES_PH",
    PITCAIRN_PN = <any>"PITCAIRN_PN",
    POLAND_PL = <any>"POLAND_PL",
    PORTUGAL_PT = <any>"PORTUGAL_PT",
    PORTUGAL_PT_20 = <any>"PORTUGAL_PT_20",
    PORTUGAL_PT_30 = <any>"PORTUGAL_PT_30",
    PUERTO_RICO_PR = <any>"PUERTO_RICO_PR",
    QATAR_QA = <any>"QATAR_QA",
    REUNION_RE = <any>"REUNION_RE",
    ROMANIA_RO = <any>"ROMANIA_RO",
    RUSSIAN_FEDERATION_RU = <any>"RUSSIAN_FEDERATION_RU",
    RWANDA_RW = <any>"RWANDA_RW",
    SAINT_BARTHELEMY_BL = <any>"SAINT_BARTHELEMY_BL",
    SAINT_HELENA_ASCENSION_AND_TRISTAN_DA_CUNHA_SH = <any>"SAINT_HELENA_ASCENSION_AND_TRISTAN_DA_CUNHA_SH",
    SAINT_KITTS_AND_NEVIS_KN = <any>"SAINT_KITTS_AND_NEVIS_KN",
    SAINT_LUCIA_LC = <any>"SAINT_LUCIA_LC",
    SAINT_MARTIN_MF = <any>"SAINT_MARTIN_MF",
    SAINT_PIERRE_AND_MIQUELON_PM = <any>"SAINT_PIERRE_AND_MIQUELON_PM",
    SAINT_VINCENT_AND_THE_GRENADINES_VC = <any>"SAINT_VINCENT_AND_THE_GRENADINES_VC",
    SAMOA_WS = <any>"SAMOA_WS",
    SAN_MARINO_SM = <any>"SAN_MARINO_SM",
    SAO_TOME_AND_PRINCIPE_ST = <any>"SAO_TOME_AND_PRINCIPE_ST",
    SAUDI_ARABIA_SA = <any>"SAUDI_ARABIA_SA",
    SENEGAL_SN = <any>"SENEGAL_SN",
    SERBIA_RS = <any>"SERBIA_RS",
    SEYCHELLES_SC = <any>"SEYCHELLES_SC",
    SIERRA_LEONE_SL = <any>"SIERRA_LEONE_SL",
    SINGAPORE_SG = <any>"SINGAPORE_SG",
    SINT_MAARTEN_SX = <any>"SINT_MAARTEN_SX",
    SLOVAKIA_SK = <any>"SLOVAKIA_SK",
    SLOVENIA_SI = <any>"SLOVENIA_SI",
    SOLOMON_ISLANDS_SB = <any>"SOLOMON_ISLANDS_SB",
    SOMALIA_SO = <any>"SOMALIA_SO",
    SOUTH_AFRICA_ZA = <any>"SOUTH_AFRICA_ZA",
    SOUTH_GEORGIA_AND_THE_SOUTH_SANDWICH_ISLANDS_GS = <any>"SOUTH_GEORGIA_AND_THE_SOUTH_SANDWICH_ISLANDS_GS",
    SOUTH_SUDAN_SS = <any>"SOUTH_SUDAN_SS",
    SPAIN_ES = <any>"SPAIN_ES",
    SRI_LANKA_LK = <any>"SRI_LANKA_LK",
    SUDAN_SD = <any>"SUDAN_SD",
    SURINAME_SR = <any>"SURINAME_SR",
    SVALBARD_AND_JAN_MAYEN_SJ = <any>"SVALBARD_AND_JAN_MAYEN_SJ",
    SWEDEN_SE = <any>"SWEDEN_SE",
    SWITZERLAND_CH = <any>"SWITZERLAND_CH",
    SYRIAN_ARAB_REPUBLIC_SY = <any>"SYRIAN_ARAB_REPUBLIC_SY",
    TAIWAN_TW = <any>"TAIWAN_TW",
    TAJIKISTAN_TJ = <any>"TAJIKISTAN_TJ",
    TANZANIA_TZ = <any>"TANZANIA_TZ",
    THAILAND_TH = <any>"THAILAND_TH",
    TIMOR_LESTE_TL = <any>"TIMOR_LESTE_TL",
    TOGO_TG = <any>"TOGO_TG",
    TOKELAU_TK = <any>"TOKELAU_TK",
    TONGA_TO = <any>"TONGA_TO",
    TRINIDAD_AND_TOBAGO_TT = <any>"TRINIDAD_AND_TOBAGO_TT",
    TUNISIA_TN = <any>"TUNISIA_TN",
    TURKEY_TR = <any>"TURKEY_TR",
    TURKMENISTAN_TM = <any>"TURKMENISTAN_TM",
    TURKS_AND_CAICOS_ISLANDS_TC = <any>"TURKS_AND_CAICOS_ISLANDS_TC",
    TUVALU_TV = <any>"TUVALU_TV",
    UGANDA_UG = <any>"UGANDA_UG",
    UKRAINE_UA = <any>"UKRAINE_UA",
    UNITED_ARAB_EMIRATES_AE = <any>"UNITED_ARAB_EMIRATES_AE",
    UNITED_ARAB_EMIRATES_AE_AJ = <any>"UNITED_ARAB_EMIRATES_AE_AJ",
    UNITED_ARAB_EMIRATES_AE_AZ = <any>"UNITED_ARAB_EMIRATES_AE_AZ",
    UNITED_ARAB_EMIRATES_AE_DU = <any>"UNITED_ARAB_EMIRATES_AE_DU",
    UNITED_ARAB_EMIRATES_AE_FU = <any>"UNITED_ARAB_EMIRATES_AE_FU",
    UNITED_ARAB_EMIRATES_AE_RK = <any>"UNITED_ARAB_EMIRATES_AE_RK",
    UNITED_ARAB_EMIRATES_AE_SH = <any>"UNITED_ARAB_EMIRATES_AE_SH",
    UNITED_ARAB_EMIRATES_AE_UQ = <any>"UNITED_ARAB_EMIRATES_AE_UQ",
    UNITED_KINGDOM_OF_GREAT_BRITAIN_AND_NORTHERN_IRELAND_GB = <any>"UNITED_KINGDOM_OF_GREAT_BRITAIN_AND_NORTHERN_IRELAND_GB",
    UNITED_KINGDOM_OF_GREAT_BRITAIN_AND_NORTHERN_IRELAND_GB_ENG = <any>"UNITED_KINGDOM_OF_GREAT_BRITAIN_AND_NORTHERN_IRELAND_GB_ENG",
    UNITED_KINGDOM_OF_GREAT_BRITAIN_AND_NORTHERN_IRELAND_GB_GBN = <any>"UNITED_KINGDOM_OF_GREAT_BRITAIN_AND_NORTHERN_IRELAND_GB_GBN",
    UNITED_KINGDOM_OF_GREAT_BRITAIN_AND_NORTHERN_IRELAND_GB_NIR = <any>"UNITED_KINGDOM_OF_GREAT_BRITAIN_AND_NORTHERN_IRELAND_GB_NIR",
    UNITED_KINGDOM_OF_GREAT_BRITAIN_AND_NORTHERN_IRELAND_GB_SCT = <any>"UNITED_KINGDOM_OF_GREAT_BRITAIN_AND_NORTHERN_IRELAND_GB_SCT",
    UNITED_KINGDOM_OF_GREAT_BRITAIN_AND_NORTHERN_IRELAND_GB_WLS = <any>"UNITED_KINGDOM_OF_GREAT_BRITAIN_AND_NORTHERN_IRELAND_GB_WLS",
    UNITED_STATES_MINOR_OUTLYING_ISLANDS_US_UM = <any>"UNITED_STATES_MINOR_OUTLYING_ISLANDS_US_UM",
    UNITED_STATES_OF_AMERICA_US_AK = <any>"UNITED_STATES_OF_AMERICA_US_AK",
    UNITED_STATES_OF_AMERICA_US_AL = <any>"UNITED_STATES_OF_AMERICA_US_AL",
    UNITED_STATES_OF_AMERICA_US_AR = <any>"UNITED_STATES_OF_AMERICA_US_AR",
    UNITED_STATES_OF_AMERICA_US_AZ = <any>"UNITED_STATES_OF_AMERICA_US_AZ",
    UNITED_STATES_OF_AMERICA_US_CA = <any>"UNITED_STATES_OF_AMERICA_US_CA",
    UNITED_STATES_OF_AMERICA_US_CO = <any>"UNITED_STATES_OF_AMERICA_US_CO",
    UNITED_STATES_OF_AMERICA_US_CT = <any>"UNITED_STATES_OF_AMERICA_US_CT",
    UNITED_STATES_OF_AMERICA_US_DC = <any>"UNITED_STATES_OF_AMERICA_US_DC",
    UNITED_STATES_OF_AMERICA_US_DE = <any>"UNITED_STATES_OF_AMERICA_US_DE",
    UNITED_STATES_OF_AMERICA_US_FL = <any>"UNITED_STATES_OF_AMERICA_US_FL",
    UNITED_STATES_OF_AMERICA_US_GA = <any>"UNITED_STATES_OF_AMERICA_US_GA",
    UNITED_STATES_OF_AMERICA_US_HI = <any>"UNITED_STATES_OF_AMERICA_US_HI",
    UNITED_STATES_OF_AMERICA_US_IA = <any>"UNITED_STATES_OF_AMERICA_US_IA",
    UNITED_STATES_OF_AMERICA_US_ID = <any>"UNITED_STATES_OF_AMERICA_US_ID",
    UNITED_STATES_OF_AMERICA_US_IL = <any>"UNITED_STATES_OF_AMERICA_US_IL",
    UNITED_STATES_OF_AMERICA_US_IN = <any>"UNITED_STATES_OF_AMERICA_US_IN",
    UNITED_STATES_OF_AMERICA_US_KS = <any>"UNITED_STATES_OF_AMERICA_US_KS",
    UNITED_STATES_OF_AMERICA_US_KY = <any>"UNITED_STATES_OF_AMERICA_US_KY",
    UNITED_STATES_OF_AMERICA_US_LA = <any>"UNITED_STATES_OF_AMERICA_US_LA",
    UNITED_STATES_OF_AMERICA_US_MA = <any>"UNITED_STATES_OF_AMERICA_US_MA",
    UNITED_STATES_OF_AMERICA_US_MD = <any>"UNITED_STATES_OF_AMERICA_US_MD",
    UNITED_STATES_OF_AMERICA_US_ME = <any>"UNITED_STATES_OF_AMERICA_US_ME",
    UNITED_STATES_OF_AMERICA_US_MI = <any>"UNITED_STATES_OF_AMERICA_US_MI",
    UNITED_STATES_OF_AMERICA_US_MN = <any>"UNITED_STATES_OF_AMERICA_US_MN",
    UNITED_STATES_OF_AMERICA_US_MO = <any>"UNITED_STATES_OF_AMERICA_US_MO",
    UNITED_STATES_OF_AMERICA_US_MS = <any>"UNITED_STATES_OF_AMERICA_US_MS",
    UNITED_STATES_OF_AMERICA_US_MT = <any>"UNITED_STATES_OF_AMERICA_US_MT",
    UNITED_STATES_OF_AMERICA_US_NC = <any>"UNITED_STATES_OF_AMERICA_US_NC",
    UNITED_STATES_OF_AMERICA_US_ND = <any>"UNITED_STATES_OF_AMERICA_US_ND",
    UNITED_STATES_OF_AMERICA_US_NE = <any>"UNITED_STATES_OF_AMERICA_US_NE",
    UNITED_STATES_OF_AMERICA_US_NH = <any>"UNITED_STATES_OF_AMERICA_US_NH",
    UNITED_STATES_OF_AMERICA_US_NJ = <any>"UNITED_STATES_OF_AMERICA_US_NJ",
    UNITED_STATES_OF_AMERICA_US_NM = <any>"UNITED_STATES_OF_AMERICA_US_NM",
    UNITED_STATES_OF_AMERICA_US_NV = <any>"UNITED_STATES_OF_AMERICA_US_NV",
    UNITED_STATES_OF_AMERICA_US_NY = <any>"UNITED_STATES_OF_AMERICA_US_NY",
    UNITED_STATES_OF_AMERICA_US_OH = <any>"UNITED_STATES_OF_AMERICA_US_OH",
    UNITED_STATES_OF_AMERICA_US_OK = <any>"UNITED_STATES_OF_AMERICA_US_OK",
    UNITED_STATES_OF_AMERICA_US_OR = <any>"UNITED_STATES_OF_AMERICA_US_OR",
    UNITED_STATES_OF_AMERICA_US_PA = <any>"UNITED_STATES_OF_AMERICA_US_PA",
    UNITED_STATES_OF_AMERICA_US_RI = <any>"UNITED_STATES_OF_AMERICA_US_RI",
    UNITED_STATES_OF_AMERICA_US_SC = <any>"UNITED_STATES_OF_AMERICA_US_SC",
    UNITED_STATES_OF_AMERICA_US_SD = <any>"UNITED_STATES_OF_AMERICA_US_SD",
    UNITED_STATES_OF_AMERICA_US_TN = <any>"UNITED_STATES_OF_AMERICA_US_TN",
    UNITED_STATES_OF_AMERICA_US_TX = <any>"UNITED_STATES_OF_AMERICA_US_TX",
    UNITED_STATES_OF_AMERICA_US_UT = <any>"UNITED_STATES_OF_AMERICA_US_UT",
    UNITED_STATES_OF_AMERICA_US_VA = <any>"UNITED_STATES_OF_AMERICA_US_VA",
    UNITED_STATES_OF_AMERICA_US_VT = <any>"UNITED_STATES_OF_AMERICA_US_VT",
    UNITED_STATES_OF_AMERICA_US_WA = <any>"UNITED_STATES_OF_AMERICA_US_WA",
    UNITED_STATES_OF_AMERICA_US_WI = <any>"UNITED_STATES_OF_AMERICA_US_WI",
    UNITED_STATES_OF_AMERICA_US_WV = <any>"UNITED_STATES_OF_AMERICA_US_WV",
    UNITED_STATES_OF_AMERICA_US_WY = <any>"UNITED_STATES_OF_AMERICA_US_WY",
    URUGUAY_UY = <any>"URUGUAY_UY",
    UZBEKISTAN_UZ = <any>"UZBEKISTAN_UZ",
    VANUATU_VU = <any>"VANUATU_VU",
    VENEZUELA_VE = <any>"VENEZUELA_VE",
    VIET_NAM_VN = <any>"VIET_NAM_VN",
    VIRGIN_ISLANDS_VG = <any>"VIRGIN_ISLANDS_VG",
    VIRGIN_ISLANDS_VI = <any>"VIRGIN_ISLANDS_VI",
    WALLIS_AND_FUTUNA_WF = <any>"WALLIS_AND_FUTUNA_WF",
    WESTERN_SAHARA_EH = <any>"WESTERN_SAHARA_EH",
    YEMEN_YE = <any>"YEMEN_YE",
    ZAMBIA_ZM = <any>"ZAMBIA_ZM",
    ZIMBABWE_ZW = <any>"ZIMBABWE_ZW",
}

export enum BomBrokerageRateBase {
    GROSS_CEDED_PREMIUM = <any>"GROSS_CEDED_PREMIUM",
    NET_CEDED_PREMIUM = <any>"NET_CEDED_PREMIUM",
}

export enum BomBrokerageType {
    FLAT = <any>"FLAT",
    RATE = <any>"RATE",
}

export enum BomClassOfBusiness {
    AGRICULTURE = <any>"AGRICULTURE",
    ALL = <any>"ALL",
    AVIATION = <any>"AVIATION",
    BANKS_AND_FINANCIAL_INSTITUTIONS = <any>"BANKS_AND_FINANCIAL_INSTITUTIONS",
    CREDIT_AND_SURETY = <any>"CREDIT_AND_SURETY",
    CYBER = <any>"CYBER",
    ENGINEERING = <any>"ENGINEERING",
    LIABILITY = <any>"LIABILITY",
    MARINE = <any>"MARINE",
    MOTOR = <any>"MOTOR",
    NON_LIFE_ACCIDENT_AND_HEALTH = <any>"NON_LIFE_ACCIDENT_AND_HEALTH",
    OIL_AND_ENERGY = <any>"OIL_AND_ENERGY",
    PROPERTY = <any>"PROPERTY",
    SPECIAL_RISKS = <any>"SPECIAL_RISKS",
    TRAVEL_AND_ASSISTANCE = <any>"TRAVEL_AND_ASSISTANCE",
}

export enum BomCompanyRole {
    BROKER = <any>"BROKER",
    CEDENT = <any>"CEDENT",
    REINSURER = <any>"REINSURER",
}

export enum BomCountry {
    AC = <any>"AC",
    AD = <any>"AD",
    AE = <any>"AE",
    AF = <any>"AF",
    AFRICA_ENGLISH_SPEAKING = <any>"AFRICA_ENGLISH_SPEAKING",
    AFRICA_FRENCH_SPEAKING = <any>"AFRICA_FRENCH_SPEAKING",
    AFRICA_NON_FRENCH_SPEAKING = <any>"AFRICA_NON_FRENCH_SPEAKING",
    AFRICA_NORTH = <any>"AFRICA_NORTH",
    AG = <any>"AG",
    AI = <any>"AI",
    AL = <any>"AL",
    AM = <any>"AM",
    AN = <any>"AN",
    AO = <any>"AO",
    AQ = <any>"AQ",
    AR = <any>"AR",
    AS = <any>"AS",
    ASIA_EXCLUDING_JAPAN = <any>"ASIA_EXCLUDING_JAPAN",
    AT = <any>"AT",
    AU = <any>"AU",
    AW = <any>"AW",
    AX = <any>"AX",
    AZ = <any>"AZ",
    BA = <any>"BA",
    BB = <any>"BB",
    BD = <any>"BD",
    BE = <any>"BE",
    BF = <any>"BF",
    BG = <any>"BG",
    BH = <any>"BH",
    BI = <any>"BI",
    BJ = <any>"BJ",
    BL = <any>"BL",
    BM = <any>"BM",
    BN = <any>"BN",
    BO = <any>"BO",
    BQ = <any>"BQ",
    BR = <any>"BR",
    BS = <any>"BS",
    BT = <any>"BT",
    BU = <any>"BU",
    BV = <any>"BV",
    BW = <any>"BW",
    BY = <any>"BY",
    BZ = <any>"BZ",
    CA = <any>"CA",
    CARIBBEAN = <any>"CARIBBEAN",
    CC = <any>"CC",
    CD = <any>"CD",
    CENTRAL_AMERICA = <any>"CENTRAL_AMERICA",
    CF = <any>"CF",
    CG = <any>"CG",
    CH = <any>"CH",
    CI = <any>"CI",
    CK = <any>"CK",
    CL = <any>"CL",
    CM = <any>"CM",
    CN = <any>"CN",
    CO = <any>"CO",
    CP = <any>"CP",
    CR = <any>"CR",
    CS = <any>"CS",
    CU = <any>"CU",
    CV = <any>"CV",
    CW = <any>"CW",
    CX = <any>"CX",
    CY = <any>"CY",
    CZ = <any>"CZ",
    DE = <any>"DE",
    DG = <any>"DG",
    DJ = <any>"DJ",
    DK = <any>"DK",
    DM = <any>"DM",
    DO = <any>"DO",
    DZ = <any>"DZ",
    EA = <any>"EA",
    EC = <any>"EC",
    EE = <any>"EE",
    EG = <any>"EG",
    EH = <any>"EH",
    ER = <any>"ER",
    ES = <any>"ES",
    ET = <any>"ET",
    EU = <any>"EU",
    EUROPE = <any>"EUROPE",
    EUROPE_EASTERN = <any>"EUROPE_EASTERN",
    EUROPE_EEC = <any>"EUROPE_EEC",
    EUROPE_EXCLUDING_EEC = <any>"EUROPE_EXCLUDING_EEC",
    EUROPE_NORTHERN_EUROPE = <any>"EUROPE_NORTHERN_EUROPE",
    EUROPE_NORTH_EASTERN_EUROPE = <any>"EUROPE_NORTH_EASTERN_EUROPE",
    EUROPE_SOUTH_EUROPE = <any>"EUROPE_SOUTH_EUROPE",
    EZ = <any>"EZ",
    FAR_EAST = <any>"FAR_EAST",
    FI = <any>"FI",
    FJ = <any>"FJ",
    FK = <any>"FK",
    FM = <any>"FM",
    FO = <any>"FO",
    FR = <any>"FR",
    FX = <any>"FX",
    GA = <any>"GA",
    GB = <any>"GB",
    GD = <any>"GD",
    GE = <any>"GE",
    GF = <any>"GF",
    GG = <any>"GG",
    GH = <any>"GH",
    GI = <any>"GI",
    GL = <any>"GL",
    GM = <any>"GM",
    GN = <any>"GN",
    GP = <any>"GP",
    GQ = <any>"GQ",
    GR = <any>"GR",
    GS = <any>"GS",
    GT = <any>"GT",
    GU = <any>"GU",
    GW = <any>"GW",
    GY = <any>"GY",
    HK = <any>"HK",
    HM = <any>"HM",
    HN = <any>"HN",
    HR = <any>"HR",
    HT = <any>"HT",
    HU = <any>"HU",
    IC = <any>"IC",
    ID = <any>"ID",
    IE = <any>"IE",
    IL = <any>"IL",
    IM = <any>"IM",
    IN = <any>"IN",
    INDIAN_SUB_CONTINENT = <any>"INDIAN_SUB_CONTINENT",
    IO = <any>"IO",
    IQ = <any>"IQ",
    IR = <any>"IR",
    IS = <any>"IS",
    IT = <any>"IT",
    JE = <any>"JE",
    JM = <any>"JM",
    JO = <any>"JO",
    JP = <any>"JP",
    KE = <any>"KE",
    KG = <any>"KG",
    KH = <any>"KH",
    KI = <any>"KI",
    KM = <any>"KM",
    KN = <any>"KN",
    KP = <any>"KP",
    KR = <any>"KR",
    KW = <any>"KW",
    KY = <any>"KY",
    KZ = <any>"KZ",
    LA = <any>"LA",
    LATIN_AMERICA = <any>"LATIN_AMERICA",
    LB = <any>"LB",
    LC = <any>"LC",
    LI = <any>"LI",
    LK = <any>"LK",
    LR = <any>"LR",
    LS = <any>"LS",
    LT = <any>"LT",
    LU = <any>"LU",
    LV = <any>"LV",
    LY = <any>"LY",
    MA = <any>"MA",
    MC = <any>"MC",
    MD = <any>"MD",
    ME = <any>"ME",
    MEXICO_CENTRAL_AMERICA = <any>"MEXICO_CENTRAL_AMERICA",
    MF = <any>"MF",
    MG = <any>"MG",
    MH = <any>"MH",
    MIDDLE_EAST = <any>"MIDDLE_EAST",
    MK = <any>"MK",
    ML = <any>"ML",
    MM = <any>"MM",
    MN = <any>"MN",
    MO = <any>"MO",
    MP = <any>"MP",
    MQ = <any>"MQ",
    MR = <any>"MR",
    MS = <any>"MS",
    MT = <any>"MT",
    MU = <any>"MU",
    MV = <any>"MV",
    MW = <any>"MW",
    MX = <any>"MX",
    MY = <any>"MY",
    MZ = <any>"MZ",
    NA = <any>"NA",
    NC = <any>"NC",
    NE = <any>"NE",
    NF = <any>"NF",
    NG = <any>"NG",
    NI = <any>"NI",
    NL = <any>"NL",
    NO = <any>"NO",
    NP = <any>"NP",
    NR = <any>"NR",
    NT = <any>"NT",
    NU = <any>"NU",
    NZ = <any>"NZ",
    OM = <any>"OM",
    PA = <any>"PA",
    PE = <any>"PE",
    PF = <any>"PF",
    PG = <any>"PG",
    PH = <any>"PH",
    PK = <any>"PK",
    PL = <any>"PL",
    PM = <any>"PM",
    PN = <any>"PN",
    PR = <any>"PR",
    PS = <any>"PS",
    PT = <any>"PT",
    PW = <any>"PW",
    PY = <any>"PY",
    QA = <any>"QA",
    RE = <any>"RE",
    RO = <any>"RO",
    RS = <any>"RS",
    RU = <any>"RU",
    RW = <any>"RW",
    SA = <any>"SA",
    SB = <any>"SB",
    SC = <any>"SC",
    SD = <any>"SD",
    SE = <any>"SE",
    SG = <any>"SG",
    SH = <any>"SH",
    SI = <any>"SI",
    SJ = <any>"SJ",
    SK = <any>"SK",
    SL = <any>"SL",
    SM = <any>"SM",
    SN = <any>"SN",
    SO = <any>"SO",
    SR = <any>"SR",
    SS = <any>"SS",
    ST = <any>"ST",
    SU = <any>"SU",
    SV = <any>"SV",
    SX = <any>"SX",
    SY = <any>"SY",
    SZ = <any>"SZ",
    TA = <any>"TA",
    TC = <any>"TC",
    TD = <any>"TD",
    TF = <any>"TF",
    TG = <any>"TG",
    TH = <any>"TH",
    TJ = <any>"TJ",
    TK = <any>"TK",
    TL = <any>"TL",
    TM = <any>"TM",
    TN = <any>"TN",
    TO = <any>"TO",
    TP = <any>"TP",
    TR = <any>"TR",
    TT = <any>"TT",
    TV = <any>"TV",
    TW = <any>"TW",
    TZ = <any>"TZ",
    UA = <any>"UA",
    UG = <any>"UG",
    UM = <any>"UM",
    UNDEFINED = <any>"UNDEFINED",
    US = <any>"US",
    USA_MIDWEST = <any>"USA_MIDWEST",
    USA_MID_ATLANTIC = <any>"USA_MID_ATLANTIC",
    USA_NORTHEAST = <any>"USA_NORTHEAST",
    USA_SOUTHEAST = <any>"USA_SOUTHEAST",
    USA_SOUTH_CENTRAL = <any>"USA_SOUTH_CENTRAL",
    UY = <any>"UY",
    UZ = <any>"UZ",
    VA = <any>"VA",
    VC = <any>"VC",
    VE = <any>"VE",
    VG = <any>"VG",
    VI = <any>"VI",
    VN = <any>"VN",
    VU = <any>"VU",
    WF = <any>"WF",
    WS = <any>"WS",
    XK = <any>"XK",
    YE = <any>"YE",
    YT = <any>"YT",
    YU = <any>"YU",
    ZA = <any>"ZA",
    ZM = <any>"ZM",
    ZR = <any>"ZR",
    ZW = <any>"ZW",
}

export enum BomFlatPremium {
    FLAT = <any>"FLAT",
    RATE = <any>"RATE",
    ROL = <any>"ROL",
    UNSPECIFIED = <any>"UNSPECIFIED",
    VARIABLE_RATE = <any>"VARIABLE_RATE",
}

export enum BomParticipantRole {
    BROKER = <any>"BROKER",
    BUREAU_LEADER = <any>"BUREAU_LEADER",
    CEDENT = <any>"CEDENT",
    CO_AGREEMENT_PARTY = <any>"CO_AGREEMENT_PARTY",
    REINSURER = <any>"REINSURER",
    SLIP_LEADER = <any>"SLIP_LEADER",
}

export enum BomPremiumIncomeType {
    EARNED_PREMIUM = <any>"EARNED_PREMIUM",
    GROSS_NET_EARNED_PREMIUM = <any>"GROSS_NET_EARNED_PREMIUM",
    GROSS_NET_WRITTEN_PREMIUM = <any>"GROSS_NET_WRITTEN_PREMIUM",
    GROSS_WRITTEN_PREMIUM = <any>"GROSS_WRITTEN_PREMIUM",
    NET_WRITTEN_PREMIUM = <any>"NET_WRITTEN_PREMIUM",
}

export enum BomReinsuranceBasis {
    ACCIDENT_YEAR = <any>"ACCIDENT_YEAR",
    ACCOUNTING_YEAR = <any>"ACCOUNTING_YEAR",
    UNDERWRITING_YEAR = <any>"UNDERWRITING_YEAR",
}

export enum BomReinsuranceContractCommand {
    ABANDON_ENDORSEMENT = <any>"ABANDON_ENDORSEMENT",
    ACCEPT_ENDORSEMENT = <any>"ACCEPT_ENDORSEMENT",
    APPROVE_FOR_FIRM_ORDER_TERMS = <any>"APPROVE_FOR_FIRM_ORDER_TERMS",
    APPROVE_FOR_SIGNING = <any>"APPROVE_FOR_SIGNING",
    APPROVE_FOR_SUBMISSION = <any>"APPROVE_FOR_SUBMISSION",
    APPROVE_SIGNING_PROPOSAL = <any>"APPROVE_SIGNING_PROPOSAL",
    BACK_TO_FIRM_ORDER_TERMS = <any>"BACK_TO_FIRM_ORDER_TERMS",
    BACK_TO_WRITTEN_LINE = <any>"BACK_TO_WRITTEN_LINE",
    CONFIRM_SIGNING_REQUEST = <any>"CONFIRM_SIGNING_REQUEST",
    CREATE_CONTRACT = <any>"CREATE_CONTRACT",
    CREATE_SIGNING_DRAFT = <any>"CREATE_SIGNING_DRAFT",
    CREATE_SIGNING_REQUEST = <any>"CREATE_SIGNING_REQUEST",
    DECLINE_CONTRACT = <any>"DECLINE_CONTRACT",
    DECLINE_SIGNING_PROPOSAL = <any>"DECLINE_SIGNING_PROPOSAL",
    DELETE_CONTRACT = <any>"DELETE_CONTRACT",
    MARK_AS_DONE = <any>"MARK_AS_DONE",
    NEGOTIATE_QUOTE = <any>"NEGOTIATE_QUOTE",
    NEGOTIATE_WRITTEN_LINE = <any>"NEGOTIATE_WRITTEN_LINE",
    REJECT_ENDORSEMENT = <any>"REJECT_ENDORSEMENT",
    REJECT_SIGNING_REQUEST = <any>"REJECT_SIGNING_REQUEST",
    REQUEST_FOR_QUOTE = <any>"REQUEST_FOR_QUOTE",
    REQUEST_FOR_SIGNED_APPROVAL = <any>"REQUEST_FOR_SIGNED_APPROVAL",
    REQUEST_FOR_WRITTEN_APPROVAL = <any>"REQUEST_FOR_WRITTEN_APPROVAL",
    ROLLBACK_CONTRACT = <any>"ROLLBACK_CONTRACT",
    ROLLBACK_TO_SIGNING_DRAFT = <any>"ROLLBACK_TO_SIGNING_DRAFT",
    SHARE_PRELIMINARY_CONTRACT = <any>"SHARE_PRELIMINARY_CONTRACT",
    SHARE_SIGNING_DRAFT = <any>"SHARE_SIGNING_DRAFT",
    SIGN_CONTRACT = <any>"SIGN_CONTRACT",
    SUBMIT_ENDORSEMENT = <any>"SUBMIT_ENDORSEMENT",
    SUBMIT_FINAL_SUBJECT_PREMIUM = <any>"SUBMIT_FINAL_SUBJECT_PREMIUM",
    SUBMIT_FIRM_ORDER_TERMS = <any>"SUBMIT_FIRM_ORDER_TERMS",
    SUBMIT_QUOTE = <any>"SUBMIT_QUOTE",
    SUBMIT_SIGNING_PROPOSAL = <any>"SUBMIT_SIGNING_PROPOSAL",
    SUBMIT_TO_REINSURERS_FOR_REVIEW = <any>"SUBMIT_TO_REINSURERS_FOR_REVIEW",
    WITHDRAW_CONTRACT = <any>"WITHDRAW_CONTRACT",
    WRITE_LINE = <any>"WRITE_LINE",
}

export enum BomReinsuranceContractProcessType {
    BINDING = <any>"BINDING",
    CHANGE = <any>"CHANGE",
    QUOTATION = <any>"QUOTATION",
    SIGNING = <any>"SIGNING",
    SUBMISSION = <any>"SUBMISSION",
}

export enum BomReinsuranceContractStatus {
    APPROVED_FOR_FOT = <any>"APPROVED_FOR_FOT",
    APPROVED_FOR_SIGNING = <any>"APPROVED_FOR_SIGNING",
    APPROVED_FOR_SUBMISSION = <any>"APPROVED_FOR_SUBMISSION",
    DECLINED = <any>"DECLINED",
    DRAFT = <any>"DRAFT",
    ENDORSEMENT_ABANDONED = <any>"ENDORSEMENT_ABANDONED",
    ENDORSEMENT_ACCEPTED = <any>"ENDORSEMENT_ACCEPTED",
    ENDORSEMENT_DRAFT = <any>"ENDORSEMENT_DRAFT",
    ENDORSEMENT_REQUESTED = <any>"ENDORSEMENT_REQUESTED",
    FIRM_ORDER_TERMS = <any>"FIRM_ORDER_TERMS",
    IN_NEGOTIATION = <any>"IN_NEGOTIATION",
    PENDING_APPROVAL_FOR_SIGNED = <any>"PENDING_APPROVAL_FOR_SIGNED",
    PENDING_APPROVAL_FOR_WRITTEN = <any>"PENDING_APPROVAL_FOR_WRITTEN",
    PRELIMINARY = <any>"PRELIMINARY",
    QUOTED = <any>"QUOTED",
    QUOTE_REQUEST = <any>"QUOTE_REQUEST",
    SIGNED = <any>"SIGNED",
    SIGNING_DRAFT = <any>"SIGNING_DRAFT",
    SIGNING_PROPOSAL = <any>"SIGNING_PROPOSAL",
    SIGNING_REQUEST = <any>"SIGNING_REQUEST",
    SUBMISSION_DRAFT = <any>"SUBMISSION_DRAFT",
    WITHDRAWN = <any>"WITHDRAWN",
    WRITTEN_LINE = <any>"WRITTEN_LINE",
}

export enum BomReinsuranceContractType {
    CAT_XOL = <any>"CAT_XOL",
    QUOTA_SHARE = <any>"QUOTA_SHARE",
}

export enum BomReinsuranceTreatyStatus {
    DRAFT = <any>"DRAFT",
    ENDORSED = <any>"ENDORSED",
    ENDORSEMENT_REQUESTED = <any>"ENDORSEMENT_REQUESTED",
    FOT_PUBLISHED = <any>"FOT_PUBLISHED",
    IN_NEGOTIATION = <any>"IN_NEGOTIATION",
    IN_QUOTATION = <any>"IN_QUOTATION",
    PARTIALLY_PLACED = <any>"PARTIALLY_PLACED",
    PLACED = <any>"PLACED",
    REQUEST_FOR_INTEREST = <any>"REQUEST_FOR_INTEREST",
    REQUEST_FOR_QUOTE = <any>"REQUEST_FOR_QUOTE",
    SIGNED_LINE = <any>"SIGNED_LINE",
    SIGNING_PROPOSAL = <any>"SIGNING_PROPOSAL",
    SIGNING_REQUEST = <any>"SIGNING_REQUEST",
    WRITTEN_LINE = <any>"WRITTEN_LINE",
}

export enum BomTransactionStatus {
    DONE = <any>"DONE",
    FAILED = <any>"FAILED",
    IN_PROGRESS = <any>"IN_PROGRESS",
}

export enum BomVariableRatePremiumLoadingType {
    MULTIPLICATIVE = <any>"MULTIPLICATIVE",
}

export enum BomWebSocketEventType {
    BUSINESS_NETWORK_MEMBERSHIP = <any>"BUSINESS_NETWORK_MEMBERSHIP",
    CLAUSE = <any>"CLAUSE",
    COMPANY = <any>"COMPANY",
    CONTRACT = <any>"CONTRACT",
    CORDA_TX_EXECUTION_STATE = <any>"CORDA_TX_EXECUTION_STATE",
    INSURANCE_SERVICE_REQUEST = <any>"INSURANCE_SERVICE_REQUEST",
    LOSS_EVENT = <any>"LOSS_EVENT",
    PORTFOLIO = <any>"PORTFOLIO",
    STATE_CHANGE_RECORD = <any>"STATE_CHANGE_RECORD",
    TEST = <any>"TEST",
    TREATY = <any>"TREATY",
}

export enum BusinessNetworkMembershipType {
    FULL = <any>"FULL",
    GUEST = <any>"GUEST",
}

export enum ClaimAcceptanceStatus {
    ACCEPTED = <any>"ACCEPTED",
    CANCELLED = <any>"CANCELLED",
    CLAIM_ADVICE = <any>"CLAIM_ADVICE",
    DRAFT = <any>"DRAFT",
    QUERIED = <any>"QUERIED",
}

export enum ClaimAmountType {
    EXPENSE_RESERVES = <any>"EXPENSE_RESERVES",
    LOSS_RESERVES = <any>"LOSS_RESERVES",
    PAID_EXPENSES = <any>"PAID_EXPENSES",
    PAID_LOSSES = <any>"PAID_LOSSES",
    REINSTATEMENT_PREMIUMS = <any>"REINSTATEMENT_PREMIUMS",
    REINSTATEMENT_PREMIUMS_RESERVES = <any>"REINSTATEMENT_PREMIUMS_RESERVES",
}

export enum ClaimCommand {
    CLAIM_ADVISE = <any>"CLAIM_ADVISE",
    UPDATE = <any>"UPDATE",
}

export enum ClauseType {
    CUSTOM = <any>"CUSTOM",
    SUBJECTIVITY = <any>"SUBJECTIVITY",
}

export enum Codex1EntityTypeSelector {
    Clause = <any>"Clause",
    Company = <any>"Company",
    Contract = <any>"Contract",
    InsuranceServiceRequest = <any>"InsuranceServiceRequest",
    Portfolio = <any>"Portfolio",
    Treaty = <any>"Treaty",
}

export enum CommissionRateBase {
    DEPOSIT_PREMIUM = <any>"DEPOSIT_PREMIUM",
    GROSS_CEDED_PREMIUM = <any>"GROSS_CEDED_PREMIUM",
    GROSS_PREMIUM = <any>"GROSS_PREMIUM",
    NET_CEDED_PREMIUM = <any>"NET_CEDED_PREMIUM",
    NET_PREMIUM = <any>"NET_PREMIUM",
}

export enum Condition {
    NO_CLAIMS = <any>"NO_CLAIMS",
}

export enum ContractAction {
    ABANDON_ENDORSEMENT = <any>"ABANDON_ENDORSEMENT",
    ACCEPT_ENDORSEMENT = <any>"ACCEPT_ENDORSEMENT",
    APPROVE_FOR_FIRM_ORDER_TERMS = <any>"APPROVE_FOR_FIRM_ORDER_TERMS",
    APPROVE_FOR_SUBMISSION = <any>"APPROVE_FOR_SUBMISSION",
    BACK_TO_FIRM_ORDER_TERMS = <any>"BACK_TO_FIRM_ORDER_TERMS",
    BACK_TO_WRITTEN_LINE = <any>"BACK_TO_WRITTEN_LINE",
    DECLINE = <any>"DECLINE",
    DECLINE_QUOTE = <any>"DECLINE_QUOTE",
    NEGOTIATE_QUOTE = <any>"NEGOTIATE_QUOTE",
    NEGOTIATE_WRITTEN_LINE = <any>"NEGOTIATE_WRITTEN_LINE",
    REJECT_ENDORSEMENT = <any>"REJECT_ENDORSEMENT",
    REQUEST_FOR_QUOTE = <any>"REQUEST_FOR_QUOTE",
    REQUEST_FOR_SIGNED_APPROVAL = <any>"REQUEST_FOR_SIGNED_APPROVAL",
    REQUEST_FOR_WRITTEN_APPROVAL = <any>"REQUEST_FOR_WRITTEN_APPROVAL",
    ROLLBACK_CONTRACT = <any>"ROLLBACK_CONTRACT",
    SAVE = <any>"SAVE",
    SHARE_WITH_BROKER = <any>"SHARE_WITH_BROKER",
    SHARE_WITH_CEDENT = <any>"SHARE_WITH_CEDENT",
    SIGN_CONTRACT = <any>"SIGN_CONTRACT",
    SUBMIT = <any>"SUBMIT",
    SUBMIT_QUOTE = <any>"SUBMIT_QUOTE",
    SUBMIT_TO_REINSURER = <any>"SUBMIT_TO_REINSURER",
    SUBMIT_TO_REINSURERS_FOR_REVIEW = <any>"SUBMIT_TO_REINSURERS_FOR_REVIEW",
    WITHDRAW = <any>"WITHDRAW",
    WITHDRAW_QUOTE = <any>"WITHDRAW_QUOTE",
    WRITE_LINE = <any>"WRITE_LINE",
}

export enum ContractChangeType {
    CEDED_PORTFOLIO_UPDATE = <any>"CEDED_PORTFOLIO_UPDATE",
    FINAL_ADJUSTMENT = <any>"FINAL_ADJUSTMENT",
    GENERAL_AMENDMENT = <any>"GENERAL_AMENDMENT",
    PERIOD = <any>"PERIOD",
    REFERENCE_NUMBER = <any>"REFERENCE_NUMBER",
    SPECIAL_ACCEPTANCE = <any>"SPECIAL_ACCEPTANCE",
}

export enum ContractStatus {
    ABANDONED = <any>"ABANDONED",
    ANNULLED = <any>"ANNULLED",
    DRAFT = <any>"DRAFT",
    FULLY_EXECUTED = <any>"FULLY_EXECUTED",
}

export enum ContractVersion {
    CONSUMED = <any>"CONSUMED",
    LAST_CONSUMED = <any>"LAST_CONSUMED",
    TRANSITORY = <any>"TRANSITORY",
    UNCONSUMED = <any>"UNCONSUMED",
}

export enum DeductibleType {
    AGGREGATED = <any>"AGGREGATED",
    PER_EVENT = <any>"PER_EVENT",
    PER_LOSS = <any>"PER_LOSS",
    PER_POLICY = <any>"PER_POLICY",
    PER_RISK = <any>"PER_RISK",
}

export enum DocumentStatus {
    DELETED = <any>"DELETED",
    PUBLISHED = <any>"PUBLISHED",
    SAVED = <any>"SAVED",
}

export enum DocumentType {
    CALCULATION = <any>"CALCULATION",
    CLAIMS_REPORT = <any>"CLAIMS_REPORT",
    CLAUSES_CONDITIONS = <any>"CLAUSES_CONDITIONS",
    COMPLIANCE = <any>"COMPLIANCE",
    EXPOSURE = <any>"EXPOSURE",
    EXPOSURE_BASE = <any>"EXPOSURE_BASE",
    LOSS_HISTORY = <any>"LOSS_HISTORY",
    MODEL_REPORT = <any>"MODEL_REPORT",
    ORIGINAL_POLICY = <any>"ORIGINAL_POLICY",
    ORIGINAL_POLICY_SIGNED = <any>"ORIGINAL_POLICY_SIGNED",
    OTHER_DOCUMENT = <any>"OTHER_DOCUMENT",
    RISKS_AND_PREMIUMS_REPORT = <any>"RISKS_AND_PREMIUMS_REPORT",
    RISK_ENGINEERING_SURVEY = <any>"RISK_ENGINEERING_SURVEY",
    UNSPECIFIED = <any>"UNSPECIFIED",
    WET_SIGNED_REINSURANCE_SLIP = <any>"WET_SIGNED_REINSURANCE_SLIP",
}

export enum ElectronicAddressType {
    EMAIL = <any>"EMAIL",
    INSTAGRAM = <any>"INSTAGRAM",
    TWITTER = <any>"TWITTER",
    WWW = <any>"WWW",
}

export enum EntityTypeSelector {
    BusinessNetworkMembership = <any>"BusinessNetworkMembership",
    Claim = <any>"Claim",
    Clause = <any>"Clause",
    Company = <any>"Company",
    DraftProcess = <any>"DraftProcess",
    InsuranceContract = <any>"InsuranceContract",
    InsuranceContractBinding = <any>"InsuranceContractBinding",
    InsuranceContractChange = <any>"InsuranceContractChange",
    InsuranceContractNegotiation = <any>"InsuranceContractNegotiation",
    InsuranceContractProcess = <any>"InsuranceContractProcess",
    InsuranceContractQuotation = <any>"InsuranceContractQuotation",
    InsuranceContractSigning = <any>"InsuranceContractSigning",
    InsuranceContractSubmission = <any>"InsuranceContractSubmission",
    InsuranceProgram = <any>"InsuranceProgram",
    InsuranceServiceRequest = <any>"InsuranceServiceRequest",
    Loss = <any>"Loss",
    LossEvent = <any>"LossEvent",
    Portfolio = <any>"Portfolio",
    ReinsuranceContract = <any>"ReinsuranceContract",
    ReportingDataExchange = <any>"ReportingDataExchange",
    RiskPlacement = <any>"RiskPlacement",
    Treaty = <any>"Treaty",
}

export enum EurapcoBrokerFields {
    BROKER_ADDRESS = <any>"BROKER_ADDRESS",
    BROKER_NAME = <any>"BROKER_NAME",
    POLICY_ID_BROKER = <any>"POLICY_ID_BROKER",
}

export enum EurapcoConveyanceType {
    AIR = <any>"AIR",
    ALL = <any>"ALL",
    INLAND_WATERS = <any>"INLAND_WATERS",
    OTHER = <any>"OTHER",
    PIPELINES = <any>"PIPELINES",
    POST_COURIER = <any>"POST_COURIER",
    RAIL = <any>"RAIL",
    ROAD = <any>"ROAD",
    SEA = <any>"SEA",
}

export enum EurapcoExposureType {
    STORAGE = <any>"STORAGE",
    TRANSIT = <any>"TRANSIT",
    TURNOVER = <any>"TURNOVER",
}

export enum EurapcoInsuranceContractNegotiationProcessAction {
    ABANDON = <any>"ABANDON",
    ACCEPT = <any>"ACCEPT",
    ACCEPT_COMPLIANCE = <any>"ACCEPT_COMPLIANCE",
    CREATE = <any>"CREATE",
    DECLINE = <any>"DECLINE",
    REJECT = <any>"REJECT",
    RESET = <any>"RESET",
    SIGN = <any>"SIGN",
    UP_DOWN_SIGN = <any>"UP_DOWN_SIGN",
    WRITE_LINE = <any>"WRITE_LINE",
}

export enum EurapcoInsuranceContractNegotiationProcessStatus {
    COMPLIANCE_CHECK = <any>"COMPLIANCE_CHECK",
    DECLINED = <any>"DECLINED",
    FOT_OFFERED = <any>"FOT_OFFERED",
    SIGNED = <any>"SIGNED",
    SIGNING_REQUESTED = <any>"SIGNING_REQUESTED",
    WRITTEN_LINE_OFFERED = <any>"WRITTEN_LINE_OFFERED",
}

export enum EurapcoPolicyHolderFields {
    CHAMBER_OF_COMMERCE_NUMBER = <any>"CHAMBER_OF_COMMERCE_NUMBER",
    POLICY_HOLDER_ADDRESS = <any>"POLICY_HOLDER_ADDRESS",
    POLICY_HOLDER_NAME = <any>"POLICY_HOLDER_NAME",
}

export enum EurapcoPremiumCalculationType {
    BASED_ON_ANNUAL_TURNOVER = <any>"BASED_ON_ANNUAL_TURNOVER",
    DECLARATION = <any>"DECLARATION",
    FIXED_PREMIUM = <any>"FIXED_PREMIUM",
}

export enum EurapcoRiskPlacementProcessAction {
    ABANDON = <any>"ABANDON",
    ADD_NOTICE_DOCUMENT = <any>"ADD_NOTICE_DOCUMENT",
    COMPLETE_RISK_PLACEMENT = <any>"COMPLETE_RISK_PLACEMENT",
    SAVE = <any>"SAVE",
    SUBMIT_FOT = <any>"SUBMIT_FOT",
}

export enum EurapcoRiskPlacementProcessStatus {
    ABANDONED = <any>"ABANDONED",
    DRAFT = <any>"DRAFT",
    NEGOTIATIONS_IN_PROGRESS = <any>"NEGOTIATIONS_IN_PROGRESS",
    PLACED = <any>"PLACED",
}

export enum EurapcoServiceType {
    BROKERAGE = <any>"BROKERAGE",
    CEDING = <any>"CEDING",
    LEADER = <any>"LEADER",
    LOSS_ADJUSTMENT = <any>"LOSS_ADJUSTMENT",
}

export enum ExportFileType {
    CSV = <any>"CSV",
    EXCEL = <any>"EXCEL",
}

export enum ExposureBaseType {
    INSURED_LIMIT = <any>"INSURED_LIMIT",
    PAYROLL = <any>"PAYROLL",
    PRODUCTION_OUTPUT = <any>"PRODUCTION_OUTPUT",
    SALARIES = <any>"SALARIES",
    TOTAL_ASSETS = <any>"TOTAL_ASSETS",
    TURNOVER_SALES = <any>"TURNOVER_SALES",
    VALUES_AT_RISK = <any>"VALUES_AT_RISK",
    VALUE_OF_ITEMS = <any>"VALUE_OF_ITEMS",
}

export enum FinancialEventType {
    ADJUSTMENT_PREMIUM = <any>"ADJUSTMENT_PREMIUM",
    INSTALMENT_PREMIUM = <any>"INSTALMENT_PREMIUM",
    INSTALMENT_PREMIUM_CHANGE = <any>"INSTALMENT_PREMIUM_CHANGE",
    LOSS = <any>"LOSS",
}

export enum GrossNetAmountType {
    CEDENT_IBNR = <any>"CEDENT_IBNR",
    EXPENSE_RESERVES = <any>"EXPENSE_RESERVES",
    LOSS_RESERVES = <any>"LOSS_RESERVES",
    PAID_EXPENSES = <any>"PAID_EXPENSES",
    PAID_LOSSES = <any>"PAID_LOSSES",
}

export enum I3ptPolicyNegotiationProcessAction {
    AMEND = <any>"AMEND",
    COMPLETE = <any>"COMPLETE",
    CONFIRM = <any>"CONFIRM",
    CONFIRM_PAYMENT = <any>"CONFIRM_PAYMENT",
    PAY = <any>"PAY",
    READ = <any>"READ",
    SAVE = <any>"SAVE",
    SUBMIT = <any>"SUBMIT",
}

export enum I3ptPolicyNegotiationProcessStatus {
    AGREED = <any>"AGREED",
    ISSUED = <any>"ISSUED",
    PAYMENT_DONE_BY_BROKER = <any>"PAYMENT_DONE_BY_BROKER",
    PAYMENT_DONE_BY_CUSTOMER = <any>"PAYMENT_DONE_BY_CUSTOMER",
    PAYMENT_RECEIVED_BY_BROKER = <any>"PAYMENT_RECEIVED_BY_BROKER",
    PAYMENT_RECEIVED_BY_INSURER = <any>"PAYMENT_RECEIVED_BY_INSURER",
    PRELIMINARY_FROM_BROKER = <any>"PRELIMINARY_FROM_BROKER",
    PRELIMINARY_FROM_INSURER = <any>"PRELIMINARY_FROM_INSURER",
    PRIVATE = <any>"PRIVATE",
}

export enum InstalmentFrequency {
    ANNUAL = <any>"ANNUAL",
    MONTHLY = <any>"MONTHLY",
    ONE_TIME = <any>"ONE_TIME",
    QUARTERLY = <any>"QUARTERLY",
    SEMIANNUALLY = <any>"SEMIANNUALLY",
}

export enum InsuranceBasis {
    ACCIDENT_YEAR = <any>"ACCIDENT_YEAR",
    ACCOUNTING_YEAR = <any>"ACCOUNTING_YEAR",
    UNDERWRITING_YEAR = <any>"UNDERWRITING_YEAR",
}

export enum InsuranceContractBindingCommand {
    ACCEPT_ENDORSEMENT = <any>"ACCEPT_ENDORSEMENT",
    APPROVE_FOR_FIRM_ORDER_TERMS = <any>"APPROVE_FOR_FIRM_ORDER_TERMS",
    BACK_TO_FIRM_ORDER_TERMS = <any>"BACK_TO_FIRM_ORDER_TERMS",
    BACK_TO_WRITTEN_LINE = <any>"BACK_TO_WRITTEN_LINE",
    CONFIRM_SIGNING_REQUEST = <any>"CONFIRM_SIGNING_REQUEST",
    CREATE_SIGNING_REQUEST = <any>"CREATE_SIGNING_REQUEST",
    DECLINE_CONTRACT = <any>"DECLINE_CONTRACT",
    MARK_AS_DONE = <any>"MARK_AS_DONE",
    NEGOTIATE_WRITTEN_LINE = <any>"NEGOTIATE_WRITTEN_LINE",
    REJECT_SIGNING_REQUEST = <any>"REJECT_SIGNING_REQUEST",
    REQUEST_FOR_SIGNED_APPROVAL = <any>"REQUEST_FOR_SIGNED_APPROVAL",
    REQUEST_FOR_WRITTEN_APPROVAL = <any>"REQUEST_FOR_WRITTEN_APPROVAL",
    ROLLBACK_CONTRACT = <any>"ROLLBACK_CONTRACT",
    SHARE_PRELIMINARY_CONTRACT = <any>"SHARE_PRELIMINARY_CONTRACT",
    SIGN_CONTRACT = <any>"SIGN_CONTRACT",
    SUBMIT_FINAL_SUBJECT_PREMIUM = <any>"SUBMIT_FINAL_SUBJECT_PREMIUM",
    SUBMIT_FIRM_ORDER_TERMS = <any>"SUBMIT_FIRM_ORDER_TERMS",
    WITHDRAW_CONTRACT = <any>"WITHDRAW_CONTRACT",
    WRITE_LINE = <any>"WRITE_LINE",
}

export enum InsuranceContractChangeCommand {
    ABANDON_ENDORSEMENT = <any>"ABANDON_ENDORSEMENT",
    ACCEPT_ENDORSEMENT = <any>"ACCEPT_ENDORSEMENT",
    FINALISE_ACCEPTED_ENDORSEMENT = <any>"FINALISE_ACCEPTED_ENDORSEMENT",
    REJECT_ENDORSEMENT = <any>"REJECT_ENDORSEMENT",
    SUBMIT_ENDORSEMENT = <any>"SUBMIT_ENDORSEMENT",
}

export enum InsuranceContractProcessCommand {
    APPROVE_FOR_SUBMISSION = <any>"APPROVE_FOR_SUBMISSION",
    CREATE_CONTRACT = <any>"CREATE_CONTRACT",
    DECLINE_CONTRACT = <any>"DECLINE_CONTRACT",
    MARK_AS_DONE = <any>"MARK_AS_DONE",
    NEGOTIATE_QUOTE = <any>"NEGOTIATE_QUOTE",
    REQUEST_FOR_QUOTE = <any>"REQUEST_FOR_QUOTE",
    ROLLBACK_CONTRACT = <any>"ROLLBACK_CONTRACT",
    SHARE_PRELIMINARY_CONTRACT = <any>"SHARE_PRELIMINARY_CONTRACT",
    SUBMIT_QUOTE = <any>"SUBMIT_QUOTE",
    WITHDRAW_CONTRACT = <any>"WITHDRAW_CONTRACT",
}

export enum InsuranceContractProcessStatus {
    APPROVED_FOR_FOT = <any>"APPROVED_FOR_FOT",
    APPROVED_FOR_SIGNING = <any>"APPROVED_FOR_SIGNING",
    APPROVED_FOR_SUBMISSION = <any>"APPROVED_FOR_SUBMISSION",
    DECLINED = <any>"DECLINED",
    DRAFT = <any>"DRAFT",
    ENDORSEMENT_ABANDONED = <any>"ENDORSEMENT_ABANDONED",
    ENDORSEMENT_ACCEPTED = <any>"ENDORSEMENT_ACCEPTED",
    ENDORSEMENT_DRAFT = <any>"ENDORSEMENT_DRAFT",
    ENDORSEMENT_REQUESTED = <any>"ENDORSEMENT_REQUESTED",
    FIRM_ORDER_TERMS = <any>"FIRM_ORDER_TERMS",
    IN_NEGOTIATION = <any>"IN_NEGOTIATION",
    PENDING_APPROVAL_FOR_SIGNED = <any>"PENDING_APPROVAL_FOR_SIGNED",
    PENDING_APPROVAL_FOR_WRITTEN = <any>"PENDING_APPROVAL_FOR_WRITTEN",
    PRELIMINARY = <any>"PRELIMINARY",
    QUOTED = <any>"QUOTED",
    SIGNED = <any>"SIGNED",
    SIGNING_DRAFT = <any>"SIGNING_DRAFT",
    SIGNING_PROPOSAL = <any>"SIGNING_PROPOSAL",
    SIGNING_REQUEST = <any>"SIGNING_REQUEST",
    SUBMISSION_DRAFT = <any>"SUBMISSION_DRAFT",
    TO_BE_QUOTED = <any>"TO_BE_QUOTED",
    WITHDRAWN = <any>"WITHDRAWN",
    WRITTEN_LINE = <any>"WRITTEN_LINE",
}

export enum InsuranceContractSigningCommand {
    ACCEPT_ENDORSEMENT = <any>"ACCEPT_ENDORSEMENT",
    APPROVE_FOR_SIGNING = <any>"APPROVE_FOR_SIGNING",
    APPROVE_SIGNING_PROPOSAL = <any>"APPROVE_SIGNING_PROPOSAL",
    CREATE_SIGNING_DRAFT = <any>"CREATE_SIGNING_DRAFT",
    DECLINE_SIGNING_PROPOSAL = <any>"DECLINE_SIGNING_PROPOSAL",
    ROLLBACK_TO_SIGNING_DRAFT = <any>"ROLLBACK_TO_SIGNING_DRAFT",
    SHARE_SIGNING_DRAFT = <any>"SHARE_SIGNING_DRAFT",
    SUBMIT_FINAL_SUBJECT_PREMIUM = <any>"SUBMIT_FINAL_SUBJECT_PREMIUM",
    SUBMIT_SIGNING_PROPOSAL = <any>"SUBMIT_SIGNING_PROPOSAL",
    WITHDRAW_CONTRACT = <any>"WITHDRAW_CONTRACT",
}

export enum InsuranceContractSubmissionCommand {
    SUBMIT_TO_REINSURERS_FOR_REVIEW = <any>"SUBMIT_TO_REINSURERS_FOR_REVIEW",
    WITHDRAW_CONTRACT = <any>"WITHDRAW_CONTRACT",
}

export enum InsuranceProductTypes {
    FACULTATIVE_QUOTA_SHARE = <any>"FACULTATIVE_QUOTA_SHARE",
    STOP_LOSS = <any>"STOP_LOSS",
    SURPLUS_SHARE = <any>"SURPLUS_SHARE",
    TREATY_EXCESS_OF_LOSS = <any>"TREATY_EXCESS_OF_LOSS",
    TREATY_QUOTA_SHARE = <any>"TREATY_QUOTA_SHARE",
}

export enum InsuranceServiceRequestStatus {
    ABANDONED = <any>"ABANDONED",
    DONE = <any>"DONE",
    PENDING = <any>"PENDING",
}

export enum InsuranceServiceRequestType {
    MODEL_RISK = <any>"MODEL_RISK",
    PRIVATE = <any>"PRIVATE",
    REINSURE = <any>"REINSURE",
}

export enum InsuranceType {
    AGRICULTURE = <any>"AGRICULTURE",
    ALL = <any>"ALL",
    AVIATION = <any>"AVIATION",
    BANKS_AND_FINANCIAL_INSTITUTIONS = <any>"BANKS_AND_FINANCIAL_INSTITUTIONS",
    CREDIT_AND_SURETY = <any>"CREDIT_AND_SURETY",
    CYBER = <any>"CYBER",
    ENGINEERING = <any>"ENGINEERING",
    LIABILITY = <any>"LIABILITY",
    MARINE = <any>"MARINE",
    MOTOR = <any>"MOTOR",
    NON_LIFE_ACCIDENT_AND_HEALTH = <any>"NON_LIFE_ACCIDENT_AND_HEALTH",
    OIL_AND_ENERGY = <any>"OIL_AND_ENERGY",
    PROPERTY = <any>"PROPERTY",
    SPECIAL_RISKS = <any>"SPECIAL_RISKS",
    TRAVEL_AND_ASSISTANCE = <any>"TRAVEL_AND_ASSISTANCE",
}

export enum IntegrationType {
    BIDIRECTIONAL = <any>"BIDIRECTIONAL",
    NO_INTEGRATION = <any>"NO_INTEGRATION",
    UNIDIRECTIONAL_FROM_THEM = <any>"UNIDIRECTIONAL_FROM_THEM",
    UNIDIRECTIONAL_FROM_US = <any>"UNIDIRECTIONAL_FROM_US",
}

export enum LimitRateBase {
    PREMIUM = <any>"PREMIUM",
}

export enum LimitType {
    AGGREGATED = <any>"AGGREGATED",
    PER_EVENT = <any>"PER_EVENT",
    PER_LOSS = <any>"PER_LOSS",
    PER_POLICY = <any>"PER_POLICY",
    PER_RISK = <any>"PER_RISK",
}

export enum LossStatus {
    CLOSED = <any>"CLOSED",
    OPEN = <any>"OPEN",
}

export enum NoteType {
    DUE_DILIGENCE = <any>"DUE_DILIGENCE",
    PROOF_OF_SCREENING = <any>"PROOF_OF_SCREENING",
    UNDERWRITER = <any>"UNDERWRITER",
}

export enum PdfTemplateType {
    COVERNOTE_CUSTOMIZE = <any>"COVERNOTE_CUSTOMIZE",
    COVERNOTE_STANDARD = <any>"COVERNOTE_STANDARD",
    SUBMISSION_DATA_CUSTOMIZE = <any>"SUBMISSION_DATA_CUSTOMIZE",
    SUBMISSION_DATA_STANDARD = <any>"SUBMISSION_DATA_STANDARD",
}

export enum PremiumIncomeType {
    EARNED_PREMIUM = <any>"EARNED_PREMIUM",
    GROSS_NET_EARNED_PREMIUM = <any>"GROSS_NET_EARNED_PREMIUM",
    GROSS_NET_WRITTEN_PREMIUM = <any>"GROSS_NET_WRITTEN_PREMIUM",
    GROSS_WRITTEN_PREMIUM = <any>"GROSS_WRITTEN_PREMIUM",
    NET_WRITTEN_PREMIUM = <any>"NET_WRITTEN_PREMIUM",
}

export enum PremiumRateBase {
    LOSS_RECOVERABLE = <any>"LOSS_RECOVERABLE",
    SUBJECT_PREMIUM_INCOME = <any>"SUBJECT_PREMIUM_INCOME",
    SUM_INSURED = <any>"SUM_INSURED",
}

export enum RegisteredIdentityType {
    DUNS = <any>"DUNS",
    LEI = <any>"LEI",
    TAX = <any>"TAX",
}

export enum ReinstatementType {
    SINGLE = <any>"SINGLE",
    UNLIMITED = <any>"UNLIMITED",
}

export enum ReinsuranceOverviewDirection {
    ALL = <any>"ALL",
    INWARD = <any>"INWARD",
    OUTWARD = <any>"OUTWARD",
}

export enum ReinsuranceOverviewType {
    FACULTATIVE = <any>"FACULTATIVE",
    TREATY = <any>"TREATY",
}

export enum ReinsuranceRiskPlacementStatus {
    DECLINED = <any>"DECLINED",
    DRAFT = <any>"DRAFT",
    ENDORSED = <any>"ENDORSED",
    ENDORSEMENT_REQUESTED = <any>"ENDORSEMENT_REQUESTED",
    FOT_PUBLISHED = <any>"FOT_PUBLISHED",
    IN_NEGOTIATION = <any>"IN_NEGOTIATION",
    IN_QUOTATION = <any>"IN_QUOTATION",
    PARTIALLY_PLACED = <any>"PARTIALLY_PLACED",
    PLACED = <any>"PLACED",
    REQUEST_FOR_INTEREST = <any>"REQUEST_FOR_INTEREST",
    REQUEST_FOR_QUOTE = <any>"REQUEST_FOR_QUOTE",
    SIGNED_LINE = <any>"SIGNED_LINE",
    WITHDRAWN = <any>"WITHDRAWN",
    WRITTEN_LINE = <any>"WRITTEN_LINE",
}

export enum ReinsuranceStrategy {
    AS_PER_ORIGINAL_POLICY = <any>"AS_PER_ORIGINAL_POLICY",
    AS_PER_SPECIFIC_TERMS = <any>"AS_PER_SPECIFIC_TERMS",
}

export enum RiskPlacementAction {
    INVOLVE_BROKER = <any>"INVOLVE_BROKER",
    INVOLVE_REINSURER = <any>"INVOLVE_REINSURER",
    SAVE = <any>"SAVE",
    SAVE_AND_SHARE = <any>"SAVE_AND_SHARE",
}

export enum ShareType {
    ASSIGNED = <any>"ASSIGNED",
    MAXIMUM = <any>"MAXIMUM",
    MINIMUM = <any>"MINIMUM",
    RETENTION = <any>"RETENTION",
}

export enum SortDirection {
    ASC = <any>"ASC",
    DESC = <any>"DESC",
}

export enum TelecomType {
    FAX = <any>"FAX",
    VOICE = <any>"VOICE",
}

export enum TotalPremiumType {
    FULL = <any>"FULL",
    NONE = <any>"NONE",
    NOT_APPLICABLE = <any>"NOT_APPLICABLE",
    PARTIAL = <any>"PARTIAL",
}

export enum TransactionStatus {
    DONE = <any>"DONE",
    FAILED = <any>"FAILED",
    IN_PROGRESS = <any>"IN_PROGRESS",
}

export enum WebSocketEventType {
    AGREEMENT_REQUEST = <any>"AGREEMENT_REQUEST",
    BUSINESS_NETWORK_MEMBERSHIP = <any>"BUSINESS_NETWORK_MEMBERSHIP",
    CLAIM = <any>"CLAIM",
    CLAUSE = <any>"CLAUSE",
    COMPANY = <any>"COMPANY",
    CORDA_TX_EXECUTION_STATE = <any>"CORDA_TX_EXECUTION_STATE",
    INSURANCE_CONTRACT = <any>"INSURANCE_CONTRACT",
    INSURANCE_CONTRACT_NEGOTIATION = <any>"INSURANCE_CONTRACT_NEGOTIATION",
    INSURANCE_SERVICE_REQUEST = <any>"INSURANCE_SERVICE_REQUEST",
    LOSS = <any>"LOSS",
    LOSS_EVENT = <any>"LOSS_EVENT",
    MESSAGE_RECEIVED = <any>"MESSAGE_RECEIVED",
    POLICY_NEGOTIATION = <any>"POLICY_NEGOTIATION",
    PORTFOLIO = <any>"PORTFOLIO",
    REINSURANCE_CONTRACT = <any>"REINSURANCE_CONTRACT",
    REPORTING_DATA_EXCHANGE = <any>"REPORTING_DATA_EXCHANGE",
    RISK_PLACEMENT = <any>"RISK_PLACEMENT",
    STATE_CHANGE_RECORD = <any>"STATE_CHANGE_RECORD",
    TEST = <any>"TEST",
    TREATY = <any>"TREATY",
}


/**
 * bundle of all enums for databinding to options, radio-buttons etc.
 * usage in component:
 *   import { AllEnums, minValueValidator, maxValueValidator } from '../../models/webapi';
 *
 *   @Component({
 *       ...
 *   })
 *   export class xxxComponent implements OnInit {
 *       allEnums = AllEnums;
 *       ...
 *       ngOnInit() {
 *           this.allEnums = AllEnums.instance;
 *       }
 *   }
*/
export class AllEnums {
    private static _instance: AllEnums = new AllEnums();
    constructor() {
        if (AllEnums._instance) {
            throw new Error("Error: Instantiation failed: Use AllEnums.instance instead of new");
        }
        AllEnums._instance = this;
    }
    static get instance(): AllEnums {
        return AllEnums._instance;
    }

    AccountCommand = AccountCommand;
    AccountItemCategory = AccountItemCategory;
    AccountItemDueStatus = AccountItemDueStatus;
    AccountItemSortProperty = AccountItemSortProperty;
    AccountItemStatus = AccountItemStatus;
    AccountItemType = AccountItemType;
    AuditDiffEntryType = AuditDiffEntryType;
    BomApplicableLawCode = BomApplicableLawCode;
    BomBrokerageRateBase = BomBrokerageRateBase;
    BomBrokerageType = BomBrokerageType;
    BomClassOfBusiness = BomClassOfBusiness;
    BomCompanyRole = BomCompanyRole;
    BomCountry = BomCountry;
    BomFlatPremium = BomFlatPremium;
    BomParticipantRole = BomParticipantRole;
    BomPremiumIncomeType = BomPremiumIncomeType;
    BomReinsuranceBasis = BomReinsuranceBasis;
    BomReinsuranceContractCommand = BomReinsuranceContractCommand;
    BomReinsuranceContractProcessType = BomReinsuranceContractProcessType;
    BomReinsuranceContractStatus = BomReinsuranceContractStatus;
    BomReinsuranceContractType = BomReinsuranceContractType;
    BomReinsuranceTreatyStatus = BomReinsuranceTreatyStatus;
    BomTransactionStatus = BomTransactionStatus;
    BomVariableRatePremiumLoadingType = BomVariableRatePremiumLoadingType;
    BomWebSocketEventType = BomWebSocketEventType;
    BusinessNetworkMembershipType = BusinessNetworkMembershipType;
    ClaimAcceptanceStatus = ClaimAcceptanceStatus;
    ClaimAmountType = ClaimAmountType;
    ClaimCommand = ClaimCommand;
    ClauseType = ClauseType;
    Codex1EntityTypeSelector = Codex1EntityTypeSelector;
    CommissionRateBase = CommissionRateBase;
    Condition = Condition;
    ContractAction = ContractAction;
    ContractChangeType = ContractChangeType;
    ContractStatus = ContractStatus;
    ContractVersion = ContractVersion;
    DeductibleType = DeductibleType;
    DocumentStatus = DocumentStatus;
    DocumentType = DocumentType;
    ElectronicAddressType = ElectronicAddressType;
    EntityTypeSelector = EntityTypeSelector;
    EurapcoBrokerFields = EurapcoBrokerFields;
    EurapcoConveyanceType = EurapcoConveyanceType;
    EurapcoExposureType = EurapcoExposureType;
    EurapcoInsuranceContractNegotiationProcessAction = EurapcoInsuranceContractNegotiationProcessAction;
    EurapcoInsuranceContractNegotiationProcessStatus = EurapcoInsuranceContractNegotiationProcessStatus;
    EurapcoPolicyHolderFields = EurapcoPolicyHolderFields;
    EurapcoPremiumCalculationType = EurapcoPremiumCalculationType;
    EurapcoRiskPlacementProcessAction = EurapcoRiskPlacementProcessAction;
    EurapcoRiskPlacementProcessStatus = EurapcoRiskPlacementProcessStatus;
    EurapcoServiceType = EurapcoServiceType;
    ExportFileType = ExportFileType;
    ExposureBaseType = ExposureBaseType;
    FinancialEventType = FinancialEventType;
    GrossNetAmountType = GrossNetAmountType;
    I3ptPolicyNegotiationProcessAction = I3ptPolicyNegotiationProcessAction;
    I3ptPolicyNegotiationProcessStatus = I3ptPolicyNegotiationProcessStatus;
    InstalmentFrequency = InstalmentFrequency;
    InsuranceBasis = InsuranceBasis;
    InsuranceContractBindingCommand = InsuranceContractBindingCommand;
    InsuranceContractChangeCommand = InsuranceContractChangeCommand;
    InsuranceContractProcessCommand = InsuranceContractProcessCommand;
    InsuranceContractProcessStatus = InsuranceContractProcessStatus;
    InsuranceContractSigningCommand = InsuranceContractSigningCommand;
    InsuranceContractSubmissionCommand = InsuranceContractSubmissionCommand;
    InsuranceProductTypes = InsuranceProductTypes;
    InsuranceServiceRequestStatus = InsuranceServiceRequestStatus;
    InsuranceServiceRequestType = InsuranceServiceRequestType;
    InsuranceType = InsuranceType;
    IntegrationType = IntegrationType;
    LimitRateBase = LimitRateBase;
    LimitType = LimitType;
    LossStatus = LossStatus;
    NoteType = NoteType;
    PdfTemplateType = PdfTemplateType;
    PremiumIncomeType = PremiumIncomeType;
    PremiumRateBase = PremiumRateBase;
    RegisteredIdentityType = RegisteredIdentityType;
    ReinstatementType = ReinstatementType;
    ReinsuranceOverviewDirection = ReinsuranceOverviewDirection;
    ReinsuranceOverviewType = ReinsuranceOverviewType;
    ReinsuranceRiskPlacementStatus = ReinsuranceRiskPlacementStatus;
    ReinsuranceStrategy = ReinsuranceStrategy;
    RiskPlacementAction = RiskPlacementAction;
    ShareType = ShareType;
    SortDirection = SortDirection;
    TelecomType = TelecomType;
    TotalPremiumType = TotalPremiumType;
    TransactionStatus = TransactionStatus;
    WebSocketEventType = WebSocketEventType;
}

/**
* union type of all enums.
* Useful for typing params and variables in generic components.
*/
export type AllEnumsType =
    | AccountCommand
    | AccountItemCategory
    | AccountItemDueStatus
    | AccountItemSortProperty
    | AccountItemStatus
    | AccountItemType
    | AuditDiffEntryType
    | BomApplicableLawCode
    | BomBrokerageRateBase
    | BomBrokerageType
    | BomClassOfBusiness
    | BomCompanyRole
    | BomCountry
    | BomFlatPremium
    | BomParticipantRole
    | BomPremiumIncomeType
    | BomReinsuranceBasis
    | BomReinsuranceContractCommand
    | BomReinsuranceContractProcessType
    | BomReinsuranceContractStatus
    | BomReinsuranceContractType
    | BomReinsuranceTreatyStatus
    | BomTransactionStatus
    | BomVariableRatePremiumLoadingType
    | BomWebSocketEventType
    | BusinessNetworkMembershipType
    | ClaimAcceptanceStatus
    | ClaimAmountType
    | ClaimCommand
    | ClauseType
    | Codex1EntityTypeSelector
    | CommissionRateBase
    | Condition
    | ContractAction
    | ContractChangeType
    | ContractStatus
    | ContractVersion
    | DeductibleType
    | DocumentStatus
    | DocumentType
    | ElectronicAddressType
    | EntityTypeSelector
    | EurapcoBrokerFields
    | EurapcoConveyanceType
    | EurapcoExposureType
    | EurapcoInsuranceContractNegotiationProcessAction
    | EurapcoInsuranceContractNegotiationProcessStatus
    | EurapcoPolicyHolderFields
    | EurapcoPremiumCalculationType
    | EurapcoRiskPlacementProcessAction
    | EurapcoRiskPlacementProcessStatus
    | EurapcoServiceType
    | ExportFileType
    | ExposureBaseType
    | FinancialEventType
    | GrossNetAmountType
    | I3ptPolicyNegotiationProcessAction
    | I3ptPolicyNegotiationProcessStatus
    | InstalmentFrequency
    | InsuranceBasis
    | InsuranceContractBindingCommand
    | InsuranceContractChangeCommand
    | InsuranceContractProcessCommand
    | InsuranceContractProcessStatus
    | InsuranceContractSigningCommand
    | InsuranceContractSubmissionCommand
    | InsuranceProductTypes
    | InsuranceServiceRequestStatus
    | InsuranceServiceRequestType
    | InsuranceType
    | IntegrationType
    | LimitRateBase
    | LimitType
    | LossStatus
    | NoteType
    | PdfTemplateType
    | PremiumIncomeType
    | PremiumRateBase
    | RegisteredIdentityType
    | ReinstatementType
    | ReinsuranceOverviewDirection
    | ReinsuranceOverviewType
    | ReinsuranceRiskPlacementStatus
    | ReinsuranceStrategy
    | RiskPlacementAction
    | ShareType
    | SortDirection
    | TelecomType
    | TotalPremiumType
    | TransactionStatus
    | WebSocketEventType
    ;
