// swagger-ts-generator/test-gen/app/models/webapi/eurapco-audit-entry.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';

import { IProcessStatus, ProcessStatus } from './process-status.model';
import { IX500Name, X500Name } from './x-500-name.model';
import { IProcessAction, ProcessAction } from './process-action.model';

export interface IEurapcoAuditEntry {
    contractId?: string;
    contractStatus?: IProcessStatus;
    contractVersion?: number;
    date?: Date;
    triggeredBy?: IX500Name;
    workflowAction?: IProcessAction;
}


export class EurapcoAuditEntry extends BaseModel implements IEurapcoAuditEntry  {

    static CONTRACT_ID_FIELD_NAME = 'contractId';
    static CONTRACT_STATUS_FIELD_NAME = 'contractStatus';
    static CONTRACT_VERSION_FIELD_NAME = 'contractVersion';
    static DATE_FIELD_NAME = 'date';
    static TRIGGERED_BY_FIELD_NAME = 'triggeredBy';
    static WORKFLOW_ACTION_FIELD_NAME = 'workflowAction';

    contractId: string;
    contractStatus: ProcessStatus;
    contractVersion: number;
    date: Date;
    triggeredBy: X500Name;
    workflowAction: ProcessAction;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    */
    constructor(values?: Partial<IEurapcoAuditEntry>) {
        super();
        this.contractStatus = new ProcessStatus(); 
        this.triggeredBy = new X500Name(); 
        this.workflowAction = new ProcessAction(); 

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<IEurapcoAuditEntry>): void {
        if (values) {
            
            const rawValues = values
            this.contractId = this.getValue<string>(rawValues, EurapcoAuditEntry.CONTRACT_ID_FIELD_NAME);
            this.contractStatus.setValues(rawValues.contractStatus);
            this.contractVersion = this.getValue<number>(rawValues, EurapcoAuditEntry.CONTRACT_VERSION_FIELD_NAME);
            this.date = this.getValue<Date>(rawValues, EurapcoAuditEntry.DATE_FIELD_NAME);
            this.triggeredBy.setValues(rawValues.triggeredBy);
            this.workflowAction.setValues(rawValues.workflowAction);
        }
    }
}
