// swagger-ts-generator/test-gen/app/models/webapi/geographic-location.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';


export interface IGeographicLocation {
    code?: string;
    name?: string;
    typeSelector?: string;
}


export class GeographicLocation extends BaseModel implements IGeographicLocation  {
    static TYPE_SELECTOR = 'GeographicLocation';

    static CODE_FIELD_NAME = 'code';
    static NAME_FIELD_NAME = 'name';
    static TYPE_SELECTOR_FIELD_NAME = 'typeSelector';

    code: string;
    name: string;
    typeSelector: string;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    */
    constructor(values?: Partial<IGeographicLocation>) {
        super();

        this.typeSelector = GeographicLocation.TYPE_SELECTOR;

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<IGeographicLocation>): void {
        if (values) {
            
            const rawValues = values
            this.code = this.getValue<string>(rawValues, GeographicLocation.CODE_FIELD_NAME);
            this.name = this.getValue<string>(rawValues, GeographicLocation.NAME_FIELD_NAME);
            this.typeSelector = this.getValue<string>(rawValues, GeographicLocation.TYPE_SELECTOR_FIELD_NAME);
        }
    }
}
