// swagger-ts-generator/test-gen/app/models/webapi/governing-law-master-data.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';


export interface IGoverningLawMasterData {
    countryName?: string;
    isoCode?: string;
    officialName?: string;
    subDivision?: string;
    subIsoCode?: string;
    value?: string;
}


export class GoverningLawMasterData extends BaseModel implements IGoverningLawMasterData  {

    static COUNTRY_NAME_FIELD_NAME = 'countryName';
    static ISO_CODE_FIELD_NAME = 'isoCode';
    static OFFICIAL_NAME_FIELD_NAME = 'officialName';
    static SUB_DIVISION_FIELD_NAME = 'subDivision';
    static SUB_ISO_CODE_FIELD_NAME = 'subIsoCode';
    static VALUE_FIELD_NAME = 'value';

    countryName: string;
    isoCode: string;
    officialName: string;
    subDivision: string;
    subIsoCode: string;
    value: string;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    */
    constructor(values?: Partial<IGoverningLawMasterData>) {
        super();

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<IGoverningLawMasterData>): void {
        if (values) {
            
            const rawValues = values
            this.countryName = this.getValue<string>(rawValues, GoverningLawMasterData.COUNTRY_NAME_FIELD_NAME);
            this.isoCode = this.getValue<string>(rawValues, GoverningLawMasterData.ISO_CODE_FIELD_NAME);
            this.officialName = this.getValue<string>(rawValues, GoverningLawMasterData.OFFICIAL_NAME_FIELD_NAME);
            this.subDivision = this.getValue<string>(rawValues, GoverningLawMasterData.SUB_DIVISION_FIELD_NAME);
            this.subIsoCode = this.getValue<string>(rawValues, GoverningLawMasterData.SUB_ISO_CODE_FIELD_NAME);
            this.value = this.getValue<string>(rawValues, GoverningLawMasterData.VALUE_FIELD_NAME);
        }
    }
}
