// swagger-ts-generator/test-gen/app/models/webapi/i-3-pt-audit-entry.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';

import { IProcessAction, ProcessAction } from './process-action.model';
import { IProcessStatus, ProcessStatus } from './process-status.model';

export interface II3PTAuditEntry {
    action?: IProcessAction;
    date?: Date;
    policyNegotiationId?: string;
    status?: IProcessStatus;
    triggeredBy?: string;
    version?: number;
}


export class I3PTAuditEntry extends BaseModel implements II3PTAuditEntry  {

    static ACTION_FIELD_NAME = 'action';
    static DATE_FIELD_NAME = 'date';
    static POLICY_NEGOTIATION_ID_FIELD_NAME = 'policyNegotiationId';
    static STATUS_FIELD_NAME = 'status';
    static TRIGGERED_BY_FIELD_NAME = 'triggeredBy';
    static VERSION_FIELD_NAME = 'version';

    action: ProcessAction;
    date: Date;
    policyNegotiationId: string;
    status: ProcessStatus;
    triggeredBy: string;
    version: number;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    */
    constructor(values?: Partial<II3PTAuditEntry>) {
        super();
        this.action = new ProcessAction(); 
        this.status = new ProcessStatus(); 

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<II3PTAuditEntry>): void {
        if (values) {
            
            const rawValues = values
            this.action.setValues(rawValues.action);
            this.date = this.getValue<Date>(rawValues, I3PTAuditEntry.DATE_FIELD_NAME);
            this.policyNegotiationId = this.getValue<string>(rawValues, I3PTAuditEntry.POLICY_NEGOTIATION_ID_FIELD_NAME);
            this.status.setValues(rawValues.status);
            this.triggeredBy = this.getValue<string>(rawValues, I3PTAuditEntry.TRIGGERED_BY_FIELD_NAME);
            this.version = this.getValue<number>(rawValues, I3PTAuditEntry.VERSION_FIELD_NAME);
        }
    }
}
