// swagger-ts-generator/test-gen/app/models/webapi/loss-event-summary.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';

import { ILossEvent, LossEvent } from './loss-event.model';

export interface ILossEventSummary extends ILossEvent{
    nrOfOpenTasks?: number;
    nrOfUnreadMessages?: number;
}


export class LossEventSummary extends LossEvent implements ILossEventSummary  {
    static TYPE_SELECTOR = 'LossEventSummary';

    static NR_OF_OPEN_TASKS_FIELD_NAME = 'nrOfOpenTasks';
    static NR_OF_UNREAD_MESSAGES_FIELD_NAME = 'nrOfUnreadMessages';

    nrOfOpenTasks: number;
    nrOfUnreadMessages: number;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    */
    constructor(values?: Partial<ILossEventSummary>) {
        super();

        this.typeSelector = LossEventSummary.TYPE_SELECTOR;

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<ILossEventSummary>): void {
        if (values) {
            super.setValues(values)
            const rawValues = values
            this.nrOfOpenTasks = this.getValue<number>(rawValues, LossEventSummary.NR_OF_OPEN_TASKS_FIELD_NAME);
            this.nrOfUnreadMessages = this.getValue<number>(rawValues, LossEventSummary.NR_OF_UNREAD_MESSAGES_FIELD_NAME);
        }
    }
}
