// swagger-ts-generator/test-gen/app/models/webapi/loss-coverage-view.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';

import { IMoney, Money } from './money.model';

export interface ILossCoverageView {
    contractId?: string;
    contractName?: string;
    coverageId?: string;
    coverageLimit?: IMoney;
    coverageName?: string;
    coverageRemainingCapacity?: IMoney;
    lossEventId?: string;
    lossId?: string;
    reinsurerPartyId?: string;
    signedLinePercentage?: string;
    treatyId?: string;
    treatyName?: string;
}


export class LossCoverageView extends BaseModel implements ILossCoverageView  {

    static CONTRACT_ID_FIELD_NAME = 'contractId';
    static CONTRACT_NAME_FIELD_NAME = 'contractName';
    static COVERAGE_ID_FIELD_NAME = 'coverageId';
    static COVERAGE_LIMIT_FIELD_NAME = 'coverageLimit';
    static COVERAGE_NAME_FIELD_NAME = 'coverageName';
    static COVERAGE_REMAINING_CAPACITY_FIELD_NAME = 'coverageRemainingCapacity';
    static LOSS_EVENT_ID_FIELD_NAME = 'lossEventId';
    static LOSS_ID_FIELD_NAME = 'lossId';
    static REINSURER_PARTY_ID_FIELD_NAME = 'reinsurerPartyId';
    static SIGNED_LINE_PERCENTAGE_FIELD_NAME = 'signedLinePercentage';
    static TREATY_ID_FIELD_NAME = 'treatyId';
    static TREATY_NAME_FIELD_NAME = 'treatyName';

    contractId: string;
    contractName: string;
    coverageId: string;
    coverageLimit: Money;
    coverageName: string;
    coverageRemainingCapacity: Money;
    lossEventId: string;
    lossId: string;
    reinsurerPartyId: string;
    signedLinePercentage: string;
    treatyId: string;
    treatyName: string;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    */
    constructor(values?: Partial<ILossCoverageView>) {
        super();
        this.coverageLimit = new Money(); 
        this.coverageRemainingCapacity = new Money(); 

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<ILossCoverageView>): void {
        if (values) {
            
            const rawValues = values
            this.contractId = this.getValue<string>(rawValues, LossCoverageView.CONTRACT_ID_FIELD_NAME);
            this.contractName = this.getValue<string>(rawValues, LossCoverageView.CONTRACT_NAME_FIELD_NAME);
            this.coverageId = this.getValue<string>(rawValues, LossCoverageView.COVERAGE_ID_FIELD_NAME);
            this.coverageLimit.setValues(rawValues.coverageLimit);
            this.coverageName = this.getValue<string>(rawValues, LossCoverageView.COVERAGE_NAME_FIELD_NAME);
            this.coverageRemainingCapacity.setValues(rawValues.coverageRemainingCapacity);
            this.lossEventId = this.getValue<string>(rawValues, LossCoverageView.LOSS_EVENT_ID_FIELD_NAME);
            this.lossId = this.getValue<string>(rawValues, LossCoverageView.LOSS_ID_FIELD_NAME);
            this.reinsurerPartyId = this.getValue<string>(rawValues, LossCoverageView.REINSURER_PARTY_ID_FIELD_NAME);
            this.signedLinePercentage = this.getValue<string>(rawValues, LossCoverageView.SIGNED_LINE_PERCENTAGE_FIELD_NAME);
            this.treatyId = this.getValue<string>(rawValues, LossCoverageView.TREATY_ID_FIELD_NAME);
            this.treatyName = this.getValue<string>(rawValues, LossCoverageView.TREATY_NAME_FIELD_NAME);
        }
    }
}
