// swagger-ts-generator/test-gen/app/models/webapi/note.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';

import { NoteType } from './enums';

export interface INote {
    description?: string;
    owner?: string;
    type?: NoteType;
}


export class Note extends BaseModel implements INote  {

    static DESCRIPTION_FIELD_NAME = 'description';
    static OWNER_FIELD_NAME = 'owner';
    static TYPE_FIELD_NAME = 'type';

    description: string;
    owner: string;
    type: NoteType;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    */
    constructor(values?: Partial<INote>) {
        super();

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<INote>): void {
        if (values) {
            
            const rawValues = values
            this.description = this.getValue<string>(rawValues, Note.DESCRIPTION_FIELD_NAME);
            this.owner = this.getValue<string>(rawValues, Note.OWNER_FIELD_NAME);
            this.type = this.getValue<NoteType>(rawValues, Note.TYPE_FIELD_NAME);
        }
    }
}
