// swagger-ts-generator/test-gen/app/models/webapi/page-response.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';


export interface IPageResponse {
    pageIndex?: number;
    pageSize?: number;
    totalItems?: number;
}


export class PageResponse extends BaseModel implements IPageResponse  {

    static PAGE_INDEX_FIELD_NAME = 'pageIndex';
    static PAGE_SIZE_FIELD_NAME = 'pageSize';
    static TOTAL_ITEMS_FIELD_NAME = 'totalItems';

    pageIndex: number;
    pageSize: number;
    totalItems: number;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    */
    constructor(values?: Partial<IPageResponse>) {
        super();

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<IPageResponse>): void {
        if (values) {
            
            const rawValues = values
            this.pageIndex = this.getValue<number>(rawValues, PageResponse.PAGE_INDEX_FIELD_NAME);
            this.pageSize = this.getValue<number>(rawValues, PageResponse.PAGE_SIZE_FIELD_NAME);
            this.totalItems = this.getValue<number>(rawValues, PageResponse.TOTAL_ITEMS_FIELD_NAME);
        }
    }
}
