// swagger-ts-generator/test-gen/app/models/webapi/pdf-template.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';

import { PdfTemplateType } from './enums';

export interface IPdfTemplate {
    templateName?: string;
    templatePath?: string;
    templateType?: PdfTemplateType;
}


export class PdfTemplate extends BaseModel implements IPdfTemplate  {

    static TEMPLATE_NAME_FIELD_NAME = 'templateName';
    static TEMPLATE_PATH_FIELD_NAME = 'templatePath';
    static TEMPLATE_TYPE_FIELD_NAME = 'templateType';

    templateName: string;
    templatePath: string;
    templateType: PdfTemplateType;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    */
    constructor(values?: Partial<IPdfTemplate>) {
        super();

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<IPdfTemplate>): void {
        if (values) {
            
            const rawValues = values
            this.templateName = this.getValue<string>(rawValues, PdfTemplate.TEMPLATE_NAME_FIELD_NAME);
            this.templatePath = this.getValue<string>(rawValues, PdfTemplate.TEMPLATE_PATH_FIELD_NAME);
            this.templateType = this.getValue<PdfTemplateType>(rawValues, PdfTemplate.TEMPLATE_TYPE_FIELD_NAME);
        }
    }
}
