// swagger-ts-generator/test-gen/app/models/webapi/period.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';


export interface IPeriod {
    end?: Date;
    start?: Date;
    typeSelector?: string;
}


export class Period extends BaseModel implements IPeriod  {
    static TYPE_SELECTOR = 'Period';

    static END_FIELD_NAME = 'end';
    static START_FIELD_NAME = 'start';
    static TYPE_SELECTOR_FIELD_NAME = 'typeSelector';

    end: Date;
    start: Date;
    typeSelector: string;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    */
    constructor(values?: Partial<IPeriod>) {
        super();

        this.typeSelector = Period.TYPE_SELECTOR;

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<IPeriod>): void {
        if (values) {
            
            const rawValues = values
            this.end = this.getValue<Date>(rawValues, Period.END_FIELD_NAME);
            this.start = this.getValue<Date>(rawValues, Period.START_FIELD_NAME);
            this.typeSelector = this.getValue<string>(rawValues, Period.TYPE_SELECTOR_FIELD_NAME);
        }
    }
}
