// swagger-ts-generator/test-gen/app/models/webapi/protocol.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';


export interface IProtocol {
    description?: string;
    id?: string;
    name?: string;
    version?: string;
}


export class Protocol extends BaseModel implements IProtocol  {

    static DESCRIPTION_FIELD_NAME = 'description';
    static ID_FIELD_NAME = 'id';
    static NAME_FIELD_NAME = 'name';
    static VERSION_FIELD_NAME = 'version';

    description: string;
    id: string;
    name: string;
    version: string;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    */
    constructor(values?: Partial<IProtocol>) {
        super();

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<IProtocol>): void {
        if (values) {
            
            const rawValues = values
            this.description = this.getValue<string>(rawValues, Protocol.DESCRIPTION_FIELD_NAME);
            this.id = this.getValue<string>(rawValues, Protocol.ID_FIELD_NAME);
            this.name = this.getValue<string>(rawValues, Protocol.NAME_FIELD_NAME);
            this.version = this.getValue<string>(rawValues, Protocol.VERSION_FIELD_NAME);
        }
    }
}
