// swagger-ts-generator/test-gen/app/models/webapi/risk-placement-process-container.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';

import { IProcessAction, ProcessAction } from './process-action.model';
import { IReinsuranceContract, ReinsuranceContract } from './reinsurance-contract.model';
import { IPrimaryInsuranceContract, PrimaryInsuranceContract } from './primary-insurance-contract.model';
import { IRiskPlacement, RiskPlacement } from './risk-placement.model';

export interface IRiskPlacementProcessContainer {
    action?: IProcessAction;
    contract?: IReinsuranceContract;
    originalPolicy?: IPrimaryInsuranceContract;
    process?: IRiskPlacement;
}


export class RiskPlacementProcessContainer extends BaseModel implements IRiskPlacementProcessContainer  {

    static ACTION_FIELD_NAME = 'action';
    static CONTRACT_FIELD_NAME = 'contract';
    static ORIGINAL_POLICY_FIELD_NAME = 'originalPolicy';
    static PROCESS_FIELD_NAME = 'process';

    action: ProcessAction;
    contract: ReinsuranceContract;
    originalPolicy: PrimaryInsuranceContract;
    process: RiskPlacement;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    */
    constructor(values?: Partial<IRiskPlacementProcessContainer>) {
        super();
        this.action = new ProcessAction(); 
        this.contract = new ReinsuranceContract(); 
        this.originalPolicy = new PrimaryInsuranceContract(); 
        this.process = new RiskPlacement(); 

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<IRiskPlacementProcessContainer>): void {
        if (values) {
            
            const rawValues = values
            this.action.setValues(rawValues.action);
            this.contract.setValues(rawValues.contract);
            this.originalPolicy.setValues(rawValues.originalPolicy);
            this.process.setValues(rawValues.process);
        }
    }
}
