// swagger-ts-generator/test-gen/app/models/webapi/secure-identity.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';

import { IX500Name, X500Name } from './x-500-name.model';

export interface ISecureIdentity {
    companyId?: string;
    nickname?: string;
    orgPrincipal?: IX500Name;
}


export class SecureIdentity extends BaseModel implements ISecureIdentity  {

    static COMPANY_ID_FIELD_NAME = 'companyId';
    static NICKNAME_FIELD_NAME = 'nickname';
    static ORG_PRINCIPAL_FIELD_NAME = 'orgPrincipal';

    companyId: string;
    nickname: string;
    orgPrincipal: X500Name;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    */
    constructor(values?: Partial<ISecureIdentity>) {
        super();
        this.orgPrincipal = new X500Name(); 

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<ISecureIdentity>): void {
        if (values) {
            
            const rawValues = values
            this.companyId = this.getValue<string>(rawValues, SecureIdentity.COMPANY_ID_FIELD_NAME);
            this.nickname = this.getValue<string>(rawValues, SecureIdentity.NICKNAME_FIELD_NAME);
            this.orgPrincipal.setValues(rawValues.orgPrincipal);
        }
    }
}
