// swagger-ts-generator/test-gen/app/models/webapi/state-change-record.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';

import { ISecureIdentity, SecureIdentity } from './secure-identity.model';

export interface IStateChangeRecord {
    actor?: ISecureIdentity;
    correlationId?: string;
    dateTime?: Date;
    operation?: string;
}


export class StateChangeRecord extends BaseModel implements IStateChangeRecord  {

    static ACTOR_FIELD_NAME = 'actor';
    static CORRELATION_ID_FIELD_NAME = 'correlationId';
    static DATE_TIME_FIELD_NAME = 'dateTime';
    static OPERATION_FIELD_NAME = 'operation';

    actor: SecureIdentity;
    correlationId: string;
    dateTime: Date;
    operation: string;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    */
    constructor(values?: Partial<IStateChangeRecord>) {
        super();
        this.actor = new SecureIdentity(); 

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<IStateChangeRecord>): void {
        if (values) {
            
            const rawValues = values
            this.actor.setValues(rawValues.actor);
            this.correlationId = this.getValue<string>(rawValues, StateChangeRecord.CORRELATION_ID_FIELD_NAME);
            this.dateTime = this.getValue<Date>(rawValues, StateChangeRecord.DATE_TIME_FIELD_NAME);
            this.operation = this.getValue<string>(rawValues, StateChangeRecord.OPERATION_FIELD_NAME);
        }
    }
}
