// swagger-ts-generator/test-gen/app/models/webapi/sub-region-data-entry.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';

import { ICountryDataEntry, CountryDataEntry } from './country-data-entry.model';

export interface ISubRegionDataEntry {
    code?: string;
    countries?: Array<ICountryDataEntry>;
    name?: string;
}


export class SubRegionDataEntry extends BaseModel implements ISubRegionDataEntry  {

    static CODE_FIELD_NAME = 'code';
    static COUNTRIES_FIELD_NAME = 'countries';
    static NAME_FIELD_NAME = 'name';

    code: string;
    countries: Array<CountryDataEntry>;
    name: string;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    */
    constructor(values?: Partial<ISubRegionDataEntry>) {
        super();
        this.countries = new Array<CountryDataEntry>(); 

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<ISubRegionDataEntry>): void {
        if (values) {
            
            const rawValues = values
            this.code = this.getValue<string>(rawValues, SubRegionDataEntry.CODE_FIELD_NAME);
            this.fillModelArray<ICountryDataEntry>(this, SubRegionDataEntry.COUNTRIES_FIELD_NAME, rawValues.countries, CountryDataEntry, SubTypeFactory.createSubTypeInstance);
            this.name = this.getValue<string>(rawValues, SubRegionDataEntry.NAME_FIELD_NAME);
        }
    }
}
