// swagger-ts-generator/test-gen/app/models/webapi/version-info.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';


export interface IVersionInfo {
    buildDate?: Date;
    buildNumber?: string;
    version?: string;
}


export class VersionInfo extends BaseModel implements IVersionInfo  {

    static BUILD_DATE_FIELD_NAME = 'buildDate';
    static BUILD_NUMBER_FIELD_NAME = 'buildNumber';
    static VERSION_FIELD_NAME = 'version';

    buildDate: Date;
    buildNumber: string;
    version: string;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    */
    constructor(values?: Partial<IVersionInfo>) {
        super();

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<IVersionInfo>): void {
        if (values) {
            
            const rawValues = values
            this.buildDate = this.getValue<Date>(rawValues, VersionInfo.BUILD_DATE_FIELD_NAME);
            this.buildNumber = this.getValue<string>(rawValues, VersionInfo.BUILD_NUMBER_FIELD_NAME);
            this.version = this.getValue<string>(rawValues, VersionInfo.VERSION_FIELD_NAME);
        }
    }
}
