// swagger-ts-generator/test-gen/app/models/webapi/web-socket-change-notification.model.ts
/**
 * This file is generated by the SwaggerTSGenerator.
 * Do not edit.
*/
/* tslint:disable */
import { BaseModel } from './base-model';
import { SubTypeFactory } from './sub-type-factory';

import { IChangeNotification, ChangeNotification } from './change-notification.model';
import { WebSocketEventType } from './enums';

export interface IWebSocketChangeNotification {
    body?: string;
    data?: IChangeNotification;
    tag?: string;
    webSocketEventType?: WebSocketEventType;
}


export class WebSocketChangeNotification extends BaseModel implements IWebSocketChangeNotification  {

    static BODY_FIELD_NAME = 'body';
    static DATA_FIELD_NAME = 'data';
    static TAG_FIELD_NAME = 'tag';
    static WEB_SOCKET_EVENT_TYPE_FIELD_NAME = 'webSocketEventType';

    body: string;
    data: ChangeNotification;
    tag: string;
    webSocketEventType: WebSocketEventType;

    /**
     * constructor
     * @param values Can be used to set a webapi response or formValues to this newly constructed model
    */
    constructor(values?: Partial<IWebSocketChangeNotification>) {
        super();
        this.data = new ChangeNotification(); 

        if (values) {
            this.setValues(values);
        }
    }

    /**
     * set the values.
     * @param values Can be used to set a webapi response to this newly constructed model
    */
    setValues(values: Partial<IWebSocketChangeNotification>): void {
        if (values) {
            
            const rawValues = values
            this.body = this.getValue<string>(rawValues, WebSocketChangeNotification.BODY_FIELD_NAME);
            this.data.setValues(rawValues.data);
            this.tag = this.getValue<string>(rawValues, WebSocketChangeNotification.TAG_FIELD_NAME);
            this.webSocketEventType = this.getValue<WebSocketEventType>(rawValues, WebSocketChangeNotification.WEB_SOCKET_EVENT_TYPE_FIELD_NAME);
        }
    }
}
